{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f2684b1-dbad-45d5-bdb5-b83989bec7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "63b5feff-aa8a-417c-8c93-9b9d74e5dee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "pybedtools.set_bedtools_path('/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f6012705-9d34-4997-8681-c7bbcc4f008b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pybedtools' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m b \u001b[38;5;241m=\u001b[39m \u001b[43mpybedtools\u001b[49m\u001b[38;5;241m.\u001b[39mBedTool()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pybedtools' is not defined"
     ]
    }
   ],
   "source": [
    "b = pybedtools.BedTool()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1b96ffb-37e9-4408-9655-669a63004186",
   "metadata": {},
   "source": [
    "# Integration tests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed81815-5e30-4ebd-9873-0c90df4e3832",
   "metadata": {},
   "source": [
    "## Todo: \n",
    "* Make a test for single-end approach, using this bam: /tscc/projects/ps-yeolab3/ekofman/sailor2/data/Hugo-A1Aligned.sortedByCoord.out.md.subset.bam\n",
    "    * All of the reverse reads map to positive strand genes\n",
    "    * All of the forward reads map to negative strand genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf17f2fb-e302-4f21-9ca5-046685acdfd7",
   "metadata": {},
   "source": [
    "### Run tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba12c2b8-2ccb-4866-86c8-a2284fd1229f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "unstranded_pair_test\n",
      "Removing old files...\n",
      "Running tests...\n",
      "Assuming 1 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/bams/citrine435.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/unstranded_pair_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t0\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tBedgraphs:\t['CT']\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\tall\n",
      "\tNumber of intervals:\t1\n",
      "\tCores:\t1\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/unstranded_pair_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 1 subsets...\n",
      "\tContig Citrine.dna\n",
      "1 total jobs\n",
      "Reads processed:\t44\n",
      "Time to process reads in min:\t0.02141\n",
      "Read Summary:\n",
      "edited         25\n",
      "no_edits       19\n",
      "total_edits    37\n",
      "dtype: int64\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit Citrine.dna_000_0_741, 0/1...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.27s/it]\n",
      "100%|██████████| 1/1 [00:01<00:00,  1.31s/it]\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py:481: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sites_for_conversion['edit_fraction'] = sites_for_conversion['count']/sites_for_conversion['coverage']\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py:482: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sites_for_conversion['start'] = sites_for_conversion['position'] - 1\n",
      "***** WARNING: File /tmp/pybedtools.mt5ov_jc.tmp has inconsistent naming convention for record:\n",
      "Citrine.dna\t437\t439\tno_barcode_Citrine.dna_438_C_A_+\tC>A\t+\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.mt5ov_jc.tmp has inconsistent naming convention for record:\n",
      "Citrine.dna\t437\t439\tno_barcode_Citrine.dna_438_C_A_+\tC>A\t+\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    37\n",
      "filtered    37\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.022 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t37\n",
      "\tNumber of unique edit sites:\n",
      "\t14\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "14 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "2 final deduplicated SAILOR-formatted sites\n",
      "Making bedgraphs for ['CT'] conversions...\n",
      "\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Current memory usage 21.833447MB; Peak: 22.94241MB\n",
      "Time elapsed: 4.85s\n",
      "-------------------------------\n",
      "Deleting intermediate files...\n",
      "-------------------------------\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "F1R2_pair_test-single_end_mode\n",
      "Removing old files...\n",
      "Running tests...\n",
      "Python is /tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/bams/F1R2_pair.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/F1R2_pair_test-single_end_mode\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\tchr17\n",
      "\tNumber of intervals:\t16\n",
      "\tCores:\t16\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 16 subsets...\n",
      "\tContig chr17\n",
      "16 total jobs\n",
      "Reads processed:\t2\n",
      "Time to process reads in min:\t0.05137\n",
      "Read Summary:\n",
      "edited         2\n",
      "total_edits    4\n",
      "dtype: int64\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit chr17_000_0_5203591, 0/16...\n",
      "\tsplit chr17_010_52035910_57239501, 10/16...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:02<00:00,  5.37it/s]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "tests_folder=\"tests/strandedness_tests/\"\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "for t in \"unstranded_pair_test\" \"F1R2_pair_test-single_end_mode\" \"F1R2_pair_test\" \"F2R1_end_second_in_pair_test\" \"same_pos_dif_reads_test\" \"tax1bp3_chr17_3665556_read_test\" \"pair_test\"\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running tests...\"\n",
    "    bash tests/strandedness_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57192b28-b498-4f81-bdcd-0d4d14f644ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "tests_folder=\"tests/singlecell_tests/\"\n",
    "for t in \"only_5_cells_test\" \"long_read_sc_test\"\n",
    "\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running old tests...\"\n",
    "    bash tests/singlecell_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8877b85c-bb47-45e7-9ad1-05667d0fbf53",
   "metadata": {},
   "source": [
    "# Integration test automatic checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5c3ed3e-13dd-4399-924e-3d1ac17ce387",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_name_to_expectations = {\n",
    "    \"unstranded_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"Citrine.dna\",\n",
    "            \"position\": 435,\n",
    "            \"count\": 22,\n",
    "            \"coverage\": 22,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \"pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49491556,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"RPL17-C18orf32,RPL17\",\n",
    "            \"feature_strand\": \"-,-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49567494,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"LIPG\",\n",
    "            \"feature_strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \n",
    "    \"F1R2_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "    \"F1R2_pair_test-single_end_mode\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 2,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "     \"F2R1_end_second_in_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43001716,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"RPL27\"\n",
    "        }]\n",
    "    },\n",
    "    \"same_pos_dif_reads_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 83199872,\n",
    "            \"count\": 9,\n",
    "            \"coverage\": 9,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"C>G\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "    \"tax1bp3_chr17_3665556_read_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 3665556,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            #\"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "        \"only_5_cells_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000524,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>A\",\n",
    "            \"strand_conversion\": \"G>T\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GATCCCTCAGTAACGG-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 4,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"G>C\",\n",
    "            \"strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "    \"long_read_sc_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000203816-AACGTGTTGGAGAGGG-16-G\",\n",
    "            \"position\": 115807969,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"A>C\",\n",
    "            \"strand_conversion\": \"T>G\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AAGTCGTACCAGGCTC-40-C\",\n",
    "            \"position\": 115805653,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AACGTGTTGGAGAGGG-40-G\",\n",
    "            \"position\": 115807015,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 8,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        }]\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "failures = 0\n",
    "for test_name, info in test_name_to_expectations.items():\n",
    "    print(\"\\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\\nChecking results for {}\".format(test_name))\n",
    "    \n",
    "    expectations_list = info.get(\"expectations\")\n",
    "    for expectations in expectations_list:\n",
    "        print(\"\\tExpecting: {}\".format(expectations))\n",
    "    \n",
    "              \n",
    "        folder = info.get(\"folder\")\n",
    "        \n",
    "        contig = expectations.get(\"contig\")\n",
    "        barcode = expectations.get(\"barcode\", None)\n",
    "        \n",
    "        position = expectations.get(\"position\")\n",
    "        \n",
    "        final_filtered_site_info_annotated = \"{}/{}/final_filtered_site_info_annotated.tsv\".format(folder, test_name)\n",
    "        final_filtered_site_info_annotated_df = pd.read_csv(final_filtered_site_info_annotated, sep='\\t', index_col=0)\n",
    "    \n",
    "        row_of_interest = final_filtered_site_info_annotated_df[\n",
    "            (final_filtered_site_info_annotated_df['position'] == position) &\\\n",
    "            (final_filtered_site_info_annotated_df['contig'].astype(str) == contig)\n",
    "        ]\n",
    "    \n",
    "    \n",
    "        if barcode:\n",
    "            row_of_interest = row_of_interest[row_of_interest.barcode == barcode]\n",
    "    \n",
    "        failure = False\n",
    "        try:\n",
    "            assert(len(row_of_interest) == expectations.get(\"num_rows\", 1))\n",
    "        except Exception as e:\n",
    "            print(\"Num rows expected: {}, was {}\".format(expectations.get(\"num_rows\", 1), len(row_of_interest)))\n",
    "            failure = True\n",
    "            \n",
    "        for attribute in list(expectations.keys()):\n",
    "            if attribute in ['count', 'coverage', 'conversion', 'strand', 'feature_name']:\n",
    "                attribute_expectation = expectations.get(attribute)\n",
    "                try:\n",
    "                    assert(row_of_interest[attribute].iloc[0] == attribute_expectation)\n",
    "                except Exception as e:\n",
    "                    print(\"Exception: {} was {}\".format(attribute, row_of_interest[attribute].iloc[0]))\n",
    "                    failure = True\n",
    "        if not failure:\n",
    "            print(\"\\n\\t >>> {} passed! <<<\\n\".format(test_name))\n",
    "        else:\n",
    "            print(\"\\n\\t ~~~ {} FAILED! ~~~\\n\".format(test_name))\n",
    "            failures += 1\n",
    "\n",
    "print(\"There were {} failures\".format(failures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b07b328-09c2-4d4e-91c5-f3f22b041648",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fa0520-9e55-4715-ae44-bce13811802a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0bf74f0-f678-4528-ab8e-4bb96e701665",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "marine_environment",
   "language": "python",
   "name": "marine_environment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
