{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f2684b1-dbad-45d5-bdb5-b83989bec7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "63b5feff-aa8a-417c-8c93-9b9d74e5dee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "pybedtools.set_bedtools_path('/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6012705-9d34-4997-8681-c7bbcc4f008b",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = pybedtools.BedTool()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1b96ffb-37e9-4408-9655-669a63004186",
   "metadata": {},
   "source": [
    "# Integration tests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed81815-5e30-4ebd-9873-0c90df4e3832",
   "metadata": {},
   "source": [
    "## Todo: \n",
    "* Make a test for single-end approach, using this bam: /tscc/projects/ps-yeolab3/ekofman/sailor2/data/Hugo-A1Aligned.sortedByCoord.out.md.subset.bam\n",
    "    * All of the reverse reads map to positive strand genes\n",
    "    * All of the forward reads map to negative strand genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf17f2fb-e302-4f21-9ca5-046685acdfd7",
   "metadata": {},
   "source": [
    "### Run tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ba12c2b8-2ccb-4866-86c8-a2284fd1229f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "pair_test\n",
      "Removing old files...\n",
      "Running tests...\n",
      "Assuming 1 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/bams/pair_example_18_49488551_49590000.sorted.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-GRCh38-3.0.0.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/pair_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\t18\n",
      "\tNumber of intervals:\t1\n",
      "\tCores:\t1\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tTrue\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/pair_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 1 subsets...\n",
      "\tContig 18\n",
      "1 total jobs\n",
      "Reads processed:\t10\n",
      "Time to process reads in min:\t0.01962\n",
      "Read Summary:\n",
      "edited         6\n",
      "no_edits       2\n",
      "secondary      2\n",
      "total_edits    6\n",
      "dtype: int64\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit 18_000_0_80373285, 0/1...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.14s/it]\n",
      "100%|██████████| 1/1 [00:01<00:00,  1.19s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    6\n",
      "filtered    4\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.02 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t4\n",
      "\tNumber of unique edit sites:\n",
      "\t2\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "2 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "2 final deduplicated SAILOR-formatted sites\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-GRCh38-3.0.0.annotation.genes.bed...\n",
      "Current memory usage 21.320465MB; Peak: 22.430359MB\n",
      "Time elapsed: 3.12s\n",
      "++++++\n",
      "Done!\n",
      "++++++\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "tests_folder=\"tests/strandedness_tests/\"\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "#for t in \"unstranded_pair_test\" \"F1R2_pair_test-single_end_mode\" \"F1R2_pair_test\" \"F2R1_end_second_in_pair_test\" \"same_pos_dif_reads_test\" \"tax1bp3_chr17_3665556_read_test\" \"pair_test\"\n",
    "for t in \"pair_test\"\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running tests...\"\n",
    "    bash tests/strandedness_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "57192b28-b498-4f81-bdcd-0d4d14f644ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "only_5_cells_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/bams/9_3000526_only_5_cells.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/only_5_cells_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tCB\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tFalse\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\t9\n",
      "\tNumber of intervals:\t16\n",
      "\tCores:\t16\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/only_5_cells_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 16 subsets...\n",
      "\tContig 9\n",
      "16 total jobs\n",
      "Reads processed:\t31\n",
      "Time to process reads in min:\t0.1486\n",
      "Read Summary:\n",
      "edited         31\n",
      "total_edits    99\n",
      "dtype: int64\n",
      "\n",
      "Contigs processed\n",
      "\n",
      "\t['9']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig 9 has 0/4 bams generated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:08<00:00,  1.85it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t4 suffixes\n",
      "Total time to concat and write bams: 0.143 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit 9_000_0_7787195, 0/16...\n",
      "\tsplit 9_010_77871950_85659145, 10/16...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:08<00:00,  8.33s/it]\n",
      "  0%|          | 0/4 [00:07<?, ?it/s]\n",
      "multiprocessing.pool.RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/utils.py\", line 475, in get_coverage_wrapper\n",
      "    filter_stats = filter_output_df(edit_info_and_coverage_joined, filters, filtered_output_filename)\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/utils.py\", line 402, in filter_output_df\n",
      "    coverage_per_unique_position_df = pd.DataFrame(filtered_output_df.groupby(\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/core/groupby/groupby.py\", line 952, in __getattr__\n",
      "    raise AttributeError(\n",
      "AttributeError: 'DataFrameGroupBy' object has no attribute 'coverage'\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 624, in <module>\n",
      "    run(bam_filepath, \n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 347, in run\n",
      "    results, total_time, total_seconds_for_contig_df = coverage_processing(output_folder, \n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 132, in coverage_processing\n",
      "    results, total_time, total_seconds_for_contig = run_coverage_calculator(edit_info_grouped_per_contig_combined, output_folder,\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/core.py\", line 327, in run_coverage_calculator\n",
      "    for _ in p.imap_unordered(get_coverage_wrapper, coverage_counting_job_params):\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/multiprocessing/pool.py\", line 868, in next\n",
      "    raise value\n",
      "AttributeError: 'DataFrameGroupBy' object has no attribute 'coverage'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "long_read_sc_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//examples/data/LR_single_cell.md.subset.filtered.sorted.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/long_read_sc_test\n",
      "\tBarcode whitelist:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//examples/data/sc_lr_barcodes.tsv.gz\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tIB\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tFalse\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\t6\n",
      "\tNumber of intervals:\t4\n",
      "\tCores:\t16\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/long_read_sc_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "Barcodes in whitelist: 3\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 4 subsets...\n",
      "\tContig 6\n",
      "4 total jobs\n",
      "Reads processed:\t13\n",
      "Time to process reads in min:\t0.14613\n",
      "Read Summary:\n",
      "edited          3\n",
      "no_edits       10\n",
      "total_edits     3\n",
      "dtype: int64\n",
      "\n",
      "Contigs processed\n",
      "\n",
      "\t['6']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig 6 has 0/4 bams generated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4/4 [00:08<00:00,  2.12s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to concat and write bams: 0.331 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit 6_000_0_37434137, 0/4...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:19<00:00, 19.47s/it]\n",
      "  0%|          | 0/400 [00:07<?, ?it/s]\n",
      "multiprocessing.pool.RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/utils.py\", line 475, in get_coverage_wrapper\n",
      "    filter_stats = filter_output_df(edit_info_and_coverage_joined, filters, filtered_output_filename)\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/utils.py\", line 402, in filter_output_df\n",
      "    coverage_per_unique_position_df = pd.DataFrame(filtered_output_df.groupby(\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/core/groupby/groupby.py\", line 952, in __getattr__\n",
      "    raise AttributeError(\n",
      "AttributeError: 'DataFrameGroupBy' object has no attribute 'coverage'\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 624, in <module>\n",
      "    run(bam_filepath, \n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 347, in run\n",
      "    results, total_time, total_seconds_for_contig_df = coverage_processing(output_folder, \n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2//marine.py\", line 132, in coverage_processing\n",
      "    results, total_time, total_seconds_for_contig = run_coverage_calculator(edit_info_grouped_per_contig_combined, output_folder,\n",
      "  File \"/tscc/projects/ps-yeolab3/ekofman/sailor2/src/core.py\", line 327, in run_coverage_calculator\n",
      "    for _ in p.imap_unordered(get_coverage_wrapper, coverage_counting_job_params):\n",
      "  File \"/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/lib/python3.8/multiprocessing/pool.py\", line 868, in next\n",
      "    raise value\n",
      "AttributeError: 'DataFrameGroupBy' object has no attribute 'coverage'\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\npwd\\necho \"Running\"\\n\\ncd ..\\npwd\\n\\nexport MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\\nexport mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\\n\\ntests_folder=\"tests/singlecell_tests/\"\\nfor t in \"only_5_cells_test\" \"long_read_sc_test\"\\n\\ndo\\n    echo $t\\n    echo \"Removing old files...\"\\n    rm $tests_folder$t/* -r\\n\\n    echo \"Running old tests...\"\\n    bash tests/singlecell_tests/scripts/$t.sh $mypython\\n   \\ndone\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mpwd\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mecho \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mRunning\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mcd ..\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mpwd\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mexport MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mexport mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mtests_folder=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtests/singlecell_tests/\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mfor t in \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43monly_5_cells_test\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlong_read_sc_test\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdo\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    echo $t\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    echo \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mRemoving old files...\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    rm $tests_folder$t/* -r\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    echo \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mRunning old tests...\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m    bash tests/singlecell_tests/scripts/$t.sh $mypython\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m   \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mdone\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/IPython/core/interactiveshell.py:2478\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2476\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2477\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2478\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2480\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2481\u001b[0m \u001b[38;5;66;03m# when using magics with decodator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2482\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2483\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/IPython/core/magics/script.py:153\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    151\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    152\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/IPython/core/magics/script.py:305\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    301\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    302\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    303\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    304\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 305\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\npwd\\necho \"Running\"\\n\\ncd ..\\npwd\\n\\nexport MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\\nexport mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\\n\\ntests_folder=\"tests/singlecell_tests/\"\\nfor t in \"only_5_cells_test\" \"long_read_sc_test\"\\n\\ndo\\n    echo $t\\n    echo \"Removing old files...\"\\n    rm $tests_folder$t/* -r\\n\\n    echo \"Running old tests...\"\\n    bash tests/singlecell_tests/scripts/$t.sh $mypython\\n   \\ndone\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "tests_folder=\"tests/singlecell_tests/\"\n",
    "for t in \"only_5_cells_test\" \"long_read_sc_test\"\n",
    "\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running old tests...\"\n",
    "    bash tests/singlecell_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8877b85c-bb47-45e7-9ad1-05667d0fbf53",
   "metadata": {},
   "source": [
    "# Integration test automatic checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f5c3ed3e-13dd-4399-924e-3d1ac17ce387",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for unstranded_pair_test\n",
      "\tExpecting: {'contig': 'Citrine.dna', 'position': 435, 'count': 22, 'coverage': 22, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '+'}\n",
      "\n",
      "\t >>> unstranded_pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for pair_test\n",
      "\tExpecting: {'contig': '18', 'position': 49491556, 'count': 2, 'coverage': 2, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'RPL17-C18orf32,RPL17', 'feature_strand': '-,-'}\n",
      "\n",
      "\t >>> pair_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '18', 'position': 49567494, 'count': 2, 'coverage': 2, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '+', 'feature_name': 'LIPG', 'feature_strand': '+'}\n",
      "Exception: count was 3\n",
      "\n",
      "\t ~~~ pair_test FAILED! ~~~\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'BRCA1', 'feature_strand': '-'}\n",
      "Exception: count was 2\n",
      "\n",
      "\t ~~~ F1R2_pair_test FAILED! ~~~\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test-single_end_mode\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 2, 'conversion': 'G>A', 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'BRCA1', 'feature_strand': '-'}\n",
      "Exception: count was 2\n",
      "\n",
      "\t ~~~ F1R2_pair_test-single_end_mode FAILED! ~~~\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F2R1_end_second_in_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43001716, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'strand_conversion': 'G>A', 'strand': '+', 'feature_name': 'RPL27'}\n",
      "\n",
      "\t >>> F2R1_end_second_in_pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for same_pos_dif_reads_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 83199872, 'count': 9, 'coverage': 9, 'conversion': 'C>G', 'strand_conversion': 'C>G', 'strand': '+', 'feature_name': 'AC139099.2'}\n",
      "Exception: count was 14\n",
      "\n",
      "\t ~~~ same_pos_dif_reads_test FAILED! ~~~\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for tax1bp3_chr17_3665556_read_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 3665556, 'num_rows': 1, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'strand_conversion': 'G>A', 'strand': '+'}\n",
      "\n",
      "\t >>> tax1bp3_chr17_3665556_read_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for only_5_cells_test\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GGGACCTTCGAGCCAC-1', 'position': 3000524, 'num_rows': 1, 'count': 1, 'coverage': 12, 'conversion': 'C>A', 'strand_conversion': 'G>T', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GGGACCTTCGAGCCAC-1', 'position': 3000525, 'num_rows': 1, 'count': 1, 'coverage': 12, 'conversion': 'C>T', 'strand_conversion': 'G>A', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GATCCCTCAGTAACGG-1', 'position': 3000525, 'num_rows': 1, 'count': 1, 'coverage': 4, 'conversion': 'C>G', 'strand_conversion': 'G>C', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for long_read_sc_test\n",
      "\tExpecting: {'contig': '6', 'barcode': 'ENSMUST00000203816-AACGTGTTGGAGAGGG-16-G', 'position': 115807969, 'num_rows': 1, 'count': 1, 'coverage': 1, 'conversion': 'A>C', 'strand_conversion': 'T>G', 'strand': '-', 'feature_name': 'Rpl32'}\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'singlecell_tests/long_read_sc_test/final_filtered_site_info_annotated.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 210\u001b[0m\n\u001b[1;32m    207\u001b[0m position \u001b[38;5;241m=\u001b[39m expectations\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mposition\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    209\u001b[0m final_filtered_site_info_annotated \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m/final_filtered_site_info_annotated.tsv\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(folder, test_name)\n\u001b[0;32m--> 210\u001b[0m final_filtered_site_info_annotated_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfinal_filtered_site_info_annotated\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    212\u001b[0m row_of_interest \u001b[38;5;241m=\u001b[39m final_filtered_site_info_annotated_df[\n\u001b[1;32m    213\u001b[0m     (final_filtered_site_info_annotated_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mposition\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m position) \u001b[38;5;241m&\u001b[39m\\\n\u001b[1;32m    214\u001b[0m     (final_filtered_site_info_annotated_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcontig\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m==\u001b[39m contig)\n\u001b[1;32m    215\u001b[0m ]\n\u001b[1;32m    218\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m barcode:\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/envs/marine_environment/lib/python3.8/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    863\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    864\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'singlecell_tests/long_read_sc_test/final_filtered_site_info_annotated.tsv'"
     ]
    }
   ],
   "source": [
    "test_name_to_expectations = {\n",
    "    \"unstranded_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"Citrine.dna\",\n",
    "            \"position\": 435,\n",
    "            \"count\": 22,\n",
    "            \"coverage\": 22,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \"pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49491556,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"RPL17-C18orf32,RPL17\",\n",
    "            \"feature_strand\": \"-,-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49567494,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"LIPG\",\n",
    "            \"feature_strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \n",
    "    \"F1R2_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "    \"F1R2_pair_test-single_end_mode\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 2,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "     \"F2R1_end_second_in_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43001716,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"RPL27\"\n",
    "        }]\n",
    "    },\n",
    "    \"same_pos_dif_reads_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 83199872,\n",
    "            \"count\": 9,\n",
    "            \"coverage\": 9,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"C>G\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "    \"tax1bp3_chr17_3665556_read_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 3665556,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            #\"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "        \"only_5_cells_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000524,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>A\",\n",
    "            \"strand_conversion\": \"G>T\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GATCCCTCAGTAACGG-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 4,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"G>C\",\n",
    "            \"strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "    \"long_read_sc_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000203816-AACGTGTTGGAGAGGG-16-G\",\n",
    "            \"position\": 115807969,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"A>C\",\n",
    "            \"strand_conversion\": \"T>G\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AAGTCGTACCAGGCTC-40-C\",\n",
    "            \"position\": 115805653,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AACGTGTTGGAGAGGG-40-G\",\n",
    "            \"position\": 115807015,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 8,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        }]\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "failures = 0\n",
    "for test_name, info in test_name_to_expectations.items():\n",
    "    print(\"\\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\\nChecking results for {}\".format(test_name))\n",
    "    \n",
    "    expectations_list = info.get(\"expectations\")\n",
    "    for expectations in expectations_list:\n",
    "        print(\"\\tExpecting: {}\".format(expectations))\n",
    "    \n",
    "              \n",
    "        folder = info.get(\"folder\")\n",
    "        \n",
    "        contig = expectations.get(\"contig\")\n",
    "        barcode = expectations.get(\"barcode\", None)\n",
    "        \n",
    "        position = expectations.get(\"position\")\n",
    "        \n",
    "        final_filtered_site_info_annotated = \"{}/{}/final_filtered_site_info_annotated.tsv\".format(folder, test_name)\n",
    "        final_filtered_site_info_annotated_df = pd.read_csv(final_filtered_site_info_annotated, sep='\\t', index_col=0)\n",
    "    \n",
    "        row_of_interest = final_filtered_site_info_annotated_df[\n",
    "            (final_filtered_site_info_annotated_df['position'] == position) &\\\n",
    "            (final_filtered_site_info_annotated_df['contig'].astype(str) == contig)\n",
    "        ]\n",
    "    \n",
    "    \n",
    "        if barcode:\n",
    "            row_of_interest = row_of_interest[row_of_interest.barcode == barcode]\n",
    "    \n",
    "        failure = False\n",
    "        try:\n",
    "            assert(len(row_of_interest) == expectations.get(\"num_rows\", 1))\n",
    "        except Exception as e:\n",
    "            print(\"Num rows expected: {}, was {}\".format(expectations.get(\"num_rows\", 1), len(row_of_interest)))\n",
    "            failure = True\n",
    "            \n",
    "        for attribute in list(expectations.keys()):\n",
    "            if attribute in ['count', 'coverage', 'conversion', 'strand', 'feature_name']:\n",
    "                attribute_expectation = expectations.get(attribute)\n",
    "                try:\n",
    "                    assert(row_of_interest[attribute].iloc[0] == attribute_expectation)\n",
    "                except Exception as e:\n",
    "                    print(\"Exception: {} was {}\".format(attribute, row_of_interest[attribute].iloc[0]))\n",
    "                    failure = True\n",
    "        if not failure:\n",
    "            print(\"\\n\\t >>> {} passed! <<<\\n\".format(test_name))\n",
    "        else:\n",
    "            print(\"\\n\\t ~~~ {} FAILED! ~~~\\n\".format(test_name))\n",
    "            failures += 1\n",
    "\n",
    "print(\"There were {} failures\".format(failures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b07b328-09c2-4d4e-91c5-f3f22b041648",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fa0520-9e55-4715-ae44-bce13811802a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0bf74f0-f678-4528-ab8e-4bb96e701665",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "marine_environment",
   "language": "python",
   "name": "marine_environment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
