{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f2684b1-dbad-45d5-bdb5-b83989bec7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "63b5feff-aa8a-417c-8c93-9b9d74e5dee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "pybedtools.set_bedtools_path('/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6012705-9d34-4997-8681-c7bbcc4f008b",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = pybedtools.BedTool()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1b96ffb-37e9-4408-9655-669a63004186",
   "metadata": {},
   "source": [
    "# Integration tests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ed81815-5e30-4ebd-9873-0c90df4e3832",
   "metadata": {},
   "source": [
    "## Todo: \n",
    "* Make a test for single-end approach, using this bam: /tscc/projects/ps-yeolab3/ekofman/sailor2/data/Hugo-A1Aligned.sortedByCoord.out.md.subset.bam\n",
    "    * All of the reverse reads map to positive strand genes\n",
    "    * All of the forward reads map to negative strand genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf17f2fb-e302-4f21-9ca5-046685acdfd7",
   "metadata": {},
   "source": [
    "### Run tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba12c2b8-2ccb-4866-86c8-a2284fd1229f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "F1R2_pair_test-single_end_mode_sailor\n",
      "Removing old files...\n",
      "Running tests...\n",
      "Python is /tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/bams/F1R2_pair.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/F1R2_pair_test-single_end_mode_sailor\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\t['CT', 'AG']\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\tchr17\n",
      "\tNumber of intervals:\t16\n",
      "\tCores:\t16\n",
      "\tVerbose:\tTrue\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\tFor single-cell: \t6 contigs at at time\n",
      "\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Contig groups to be processed: [['chr17']]\n",
      "Each contig is being split into 16 subsets...\n",
      "contig_lengths_dict:{'chr1': 248956422, 'chr2': 242193529, 'chr3': 198295559, 'chr4': 190214555, 'chr5': 181538259, 'chr6': 170805979, 'chr7': 159345973, 'chr8': 145138636, 'chr9': 138394717, 'chr10': 133797422, 'chr11': 135086622, 'chr12': 133275309, 'chr13': 114364328, 'chr14': 107043718, 'chr15': 101991189, 'chr16': 90338345, 'chr17': 83257441, 'chr18': 80373285, 'chr19': 58617616, 'chr20': 64444167, 'chr21': 46709983, 'chr22': 50818468, 'chrX': 156040895, 'chrY': 57227415, 'chrM': 16569, 'chr1_KI270706v1_random': 175055, 'chr1_KI270707v1_random': 32032, 'chr1_KI270708v1_random': 127682, 'chr1_KI270709v1_random': 66860, 'chr1_KI270710v1_random': 40176, 'chr1_KI270711v1_random': 42210, 'chr1_KI270712v1_random': 176043, 'chr1_KI270713v1_random': 40745, 'chr1_KI270714v1_random': 41717, 'chr2_KI270715v1_random': 161471, 'chr2_KI270716v1_random': 153799, 'chr3_GL000221v1_random': 155397, 'chr4_GL000008v2_random': 209709, 'chr5_GL000208v1_random': 92689, 'chr9_KI270717v1_random': 40062, 'chr9_KI270718v1_random': 38054, 'chr9_KI270719v1_random': 176845, 'chr9_KI270720v1_random': 39050, 'chr11_KI270721v1_random': 100316, 'chr14_GL000009v2_random': 201709, 'chr14_GL000225v1_random': 211173, 'chr14_KI270722v1_random': 194050, 'chr14_GL000194v1_random': 191469, 'chr14_KI270723v1_random': 38115, 'chr14_KI270724v1_random': 39555, 'chr14_KI270725v1_random': 172810, 'chr14_KI270726v1_random': 43739, 'chr15_KI270727v1_random': 448248, 'chr16_KI270728v1_random': 1872759, 'chr17_GL000205v2_random': 185591, 'chr17_KI270729v1_random': 280839, 'chr17_KI270730v1_random': 112551, 'chr22_KI270731v1_random': 150754, 'chr22_KI270732v1_random': 41543, 'chr22_KI270733v1_random': 179772, 'chr22_KI270734v1_random': 165050, 'chr22_KI270735v1_random': 42811, 'chr22_KI270736v1_random': 181920, 'chr22_KI270737v1_random': 103838, 'chr22_KI270738v1_random': 99375, 'chr22_KI270739v1_random': 73985, 'chrY_KI270740v1_random': 37240, 'chrUn_KI270302v1': 2274, 'chrUn_KI270304v1': 2165, 'chrUn_KI270303v1': 1942, 'chrUn_KI270305v1': 1472, 'chrUn_KI270322v1': 21476, 'chrUn_KI270320v1': 4416, 'chrUn_KI270310v1': 1201, 'chrUn_KI270316v1': 1444, 'chrUn_KI270315v1': 2276, 'chrUn_KI270312v1': 998, 'chrUn_KI270311v1': 12399, 'chrUn_KI270317v1': 37690, 'chrUn_KI270412v1': 1179, 'chrUn_KI270411v1': 2646, 'chrUn_KI270414v1': 2489, 'chrUn_KI270419v1': 1029, 'chrUn_KI270418v1': 2145, 'chrUn_KI270420v1': 2321, 'chrUn_KI270424v1': 2140, 'chrUn_KI270417v1': 2043, 'chrUn_KI270422v1': 1445, 'chrUn_KI270423v1': 981, 'chrUn_KI270425v1': 1884, 'chrUn_KI270429v1': 1361, 'chrUn_KI270442v1': 392061, 'chrUn_KI270466v1': 1233, 'chrUn_KI270465v1': 1774, 'chrUn_KI270467v1': 3920, 'chrUn_KI270435v1': 92983, 'chrUn_KI270438v1': 112505, 'chrUn_KI270468v1': 4055, 'chrUn_KI270510v1': 2415, 'chrUn_KI270509v1': 2318, 'chrUn_KI270518v1': 2186, 'chrUn_KI270508v1': 1951, 'chrUn_KI270516v1': 1300, 'chrUn_KI270512v1': 22689, 'chrUn_KI270519v1': 138126, 'chrUn_KI270522v1': 5674, 'chrUn_KI270511v1': 8127, 'chrUn_KI270515v1': 6361, 'chrUn_KI270507v1': 5353, 'chrUn_KI270517v1': 3253, 'chrUn_KI270529v1': 1899, 'chrUn_KI270528v1': 2983, 'chrUn_KI270530v1': 2168, 'chrUn_KI270539v1': 993, 'chrUn_KI270538v1': 91309, 'chrUn_KI270544v1': 1202, 'chrUn_KI270548v1': 1599, 'chrUn_KI270583v1': 1400, 'chrUn_KI270587v1': 2969, 'chrUn_KI270580v1': 1553, 'chrUn_KI270581v1': 7046, 'chrUn_KI270579v1': 31033, 'chrUn_KI270589v1': 44474, 'chrUn_KI270590v1': 4685, 'chrUn_KI270584v1': 4513, 'chrUn_KI270582v1': 6504, 'chrUn_KI270588v1': 6158, 'chrUn_KI270593v1': 3041, 'chrUn_KI270591v1': 5796, 'chrUn_KI270330v1': 1652, 'chrUn_KI270329v1': 1040, 'chrUn_KI270334v1': 1368, 'chrUn_KI270333v1': 2699, 'chrUn_KI270335v1': 1048, 'chrUn_KI270338v1': 1428, 'chrUn_KI270340v1': 1428, 'chrUn_KI270336v1': 1026, 'chrUn_KI270337v1': 1121, 'chrUn_KI270363v1': 1803, 'chrUn_KI270364v1': 2855, 'chrUn_KI270362v1': 3530, 'chrUn_KI270366v1': 8320, 'chrUn_KI270378v1': 1048, 'chrUn_KI270379v1': 1045, 'chrUn_KI270389v1': 1298, 'chrUn_KI270390v1': 2387, 'chrUn_KI270387v1': 1537, 'chrUn_KI270395v1': 1143, 'chrUn_KI270396v1': 1880, 'chrUn_KI270388v1': 1216, 'chrUn_KI270394v1': 970, 'chrUn_KI270386v1': 1788, 'chrUn_KI270391v1': 1484, 'chrUn_KI270383v1': 1750, 'chrUn_KI270393v1': 1308, 'chrUn_KI270384v1': 1658, 'chrUn_KI270392v1': 971, 'chrUn_KI270381v1': 1930, 'chrUn_KI270385v1': 990, 'chrUn_KI270382v1': 4215, 'chrUn_KI270376v1': 1136, 'chrUn_KI270374v1': 2656, 'chrUn_KI270372v1': 1650, 'chrUn_KI270373v1': 1451, 'chrUn_KI270375v1': 2378, 'chrUn_KI270371v1': 2805, 'chrUn_KI270448v1': 7992, 'chrUn_KI270521v1': 7642, 'chrUn_GL000195v1': 182896, 'chrUn_GL000219v1': 179198, 'chrUn_GL000220v1': 161802, 'chrUn_GL000224v1': 179693, 'chrUn_KI270741v1': 157432, 'chrUn_GL000226v1': 15008, 'chrUn_GL000213v1': 164239, 'chrUn_KI270743v1': 210658, 'chrUn_KI270744v1': 168472, 'chrUn_KI270745v1': 41891, 'chrUn_KI270746v1': 66486, 'chrUn_KI270747v1': 198735, 'chrUn_KI270748v1': 93321, 'chrUn_KI270749v1': 158759, 'chrUn_KI270750v1': 148850, 'chrUn_KI270751v1': 150742, 'chrUn_KI270752v1': 27745, 'chrUn_KI270753v1': 62944, 'chrUn_KI270754v1': 40191, 'chrUn_KI270755v1': 36723, 'chrUn_KI270756v1': 79590, 'chrUn_KI270757v1': 71251, 'chrUn_GL000214v1': 137718, 'chrUn_KI270742v1': 186739, 'chrUn_GL000216v2': 176608, 'chrUn_GL000218v1': 161147, 'chrEBV': 171823}\n",
      "\tContig chr17\n",
      "16 total jobs\n",
      "chr17:000, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:002, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:004, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:007, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:001, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:003, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:005, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 26A20G52\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "is_proper_pair True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "VH01429:22:AACFJ5NHV:1:2211:40480:18323\t99\t#16\t43044305\t255\t1S100M\t#16\t43044312\t107\tGTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\tarray('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\t[('MD', '26A20G52'), ('PG', 'MarkDuplicates'), ('RG', 'ET_STAMPExpt_1-CTRL_S2'), ('NH', 1), ('HI', 1), ('NM', 2), ('AS', 190)]\n",
      "reverse_or_forward: -\n",
      "----------------------------\n",
      "MD tag: 26A20G52\n",
      "CIGAR string 1S100M\n",
      "Reference seq: TGCTACCAAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Aligned seq: GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities: array('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities before clipping:\n",
      " array('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "Soft clipping quality scores ...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities after clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['26', '20', '52']\n",
      "[26, 47, 100]\n",
      "Indicated reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "Fixed aligned seq:\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Finalized fixed aligned seq:\n",
      " tgctaccaagtttatttgcagtgttaCcagcacaacatttacaaaacAtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "Indicated qualities:\n",
      " 34343434343434343434263434343434343434342634343434261234343434342626343434343434343434341234341234343434343434123434343434343412343434343412343434343434343434342634342634343426343426343434342634343434\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['C', 'A']\n",
      "ref bases ['A', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['C', 'A'], ref bases: ['A', 'G']\n",
      "Getting info: C A 12 27\n",
      "Getting info: A G 12 48\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 40G3T55\n",
      "CIGAR tag 100M1S\n",
      "is_reverse True\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "is_proper_pair True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "VH01429:22:AACFJ5NHV:1:2211:40480:18323\t147\t#16\t43044312\t255\t100M1S\t#16\t43044305\t-107\tAAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\tarray('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\t[('MD', '40G3T55'), ('PG', 'MarkDuplicates'), ('RG', 'ET_STAMPExpt_1-CTRL_S2'), ('NH', 1), ('HI', 1), ('NM', 2), ('AS', 190)]\n",
      "reverse_or_forward: -\n",
      "----------------------------\n",
      "MD tag: 40G3T55\n",
      "CIGAR string 100M1S\n",
      "Reference seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Aligned seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "Qualities: array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 100), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "Qualities before clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\n",
      "Soft clipping quality scores ...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Qualities after clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34])\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['40', '3', '55']\n",
      "[40, 44, 100]\n",
      "Indicated reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "Fixed aligned seq:\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Finalized fixed aligned seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacAtatGttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "Indicated qualities:\n",
      " chr17:006, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:009, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:010, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:011, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "34343434343434263434342634343434341234343434263434343434343434343434343426343434343434341234343434343434343434343434263412343434343434343434342634343434343434343434343434343426343434263434343434343434\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['A', 'G']\n",
      "ref bases ['G', 'T']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['A', 'G'], ref bases: ['G', 'T']\n",
      "Getting info: A G 34 41\n",
      "Getting info: G T 12 45\n",
      "chr17:008, total reads: 2, counts_df:              chr17\n",
      "edited           2\n",
      "total_edits      4\n",
      "chr17:014, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:012, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:013, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:015, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "Reads processed:\t2\n",
      "Time to process reads in min:\t0.09623\n",
      "Read Summary:\n",
      "edited         2\n",
      "total_edits    4\n",
      "dtype: int64\n",
      "\n",
      "Deleting overall_label_to_list_of_contents...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit chr17_000_0_5203591, 0/16...\n",
      "\tsplit chr17_010_52035910_57239501, 10/16...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_000_0_5203591', 'chr17_001_5203591_10407182', 'chr17_002_10407182_15610773', 'chr17_003_15610773_20814364', 'chr17_004_20814364_26017955', 'chr17_005_26017955_31221546', 'chr17_006_31221546_36425137', 'chr17_007_36425137_41628728', 'chr17_008_41628728_46832319', 'chr17_009_46832319_52035910', 'chr17_010_52035910_57239501', 'chr17_011_57239501_62443092', 'chr17_012_62443092_67646683', 'chr17_013_67646683_72850274', 'chr17_014_72850274_78053865', 'chr17_015_78053865_83257456'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:05<00:00,  2.88it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:05<00:00,  3.07it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    4\n",
      "filtered    3\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.089 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t3\n",
      "\tNumber of unique edit sites:\n",
      "\t3\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "3 sites being converted to SAILOR format...\n",
      "Generating SAILOR-style bed outputs for conversion CT...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "1 final deduplicated C>T SAILOR-formatted sites\n",
      "Generating SAILOR-style bed outputs for conversion AG...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "0 final deduplicated A>G SAILOR-formatted sites\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Current memory usage 21.287284MB; Peak: 22.399873MB\n",
      "Time elapsed: 12.00s\n",
      "-------------------------------\n",
      "Deleting intermediate files...\n",
      "-------------------------------\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "F1R2_pair_test-single_end_mode\n",
      "Removing old files...\n",
      "Running tests...\n",
      "Python is /tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/bams/F1R2_pair.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/strandedness_tests/F1R2_pair_test-single_end_mode\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\t['CT']\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\tchr17\n",
      "\tNumber of intervals:\t16\n",
      "\tCores:\t16\n",
      "\tVerbose:\tTrue\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\tFor single-cell: \t6 contigs at at time\n",
      "\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Contig groups to be processed: [['chr17']]\n",
      "Each contig is being split into 16 subsets...\n",
      "contig_lengths_dict:{'chr1': 248956422, 'chr2': 242193529, 'chr3': 198295559, 'chr4': 190214555, 'chr5': 181538259, 'chr6': 170805979, 'chr7': 159345973, 'chr8': 145138636, 'chr9': 138394717, 'chr10': 133797422, 'chr11': 135086622, 'chr12': 133275309, 'chr13': 114364328, 'chr14': 107043718, 'chr15': 101991189, 'chr16': 90338345, 'chr17': 83257441, 'chr18': 80373285, 'chr19': 58617616, 'chr20': 64444167, 'chr21': 46709983, 'chr22': 50818468, 'chrX': 156040895, 'chrY': 57227415, 'chrM': 16569, 'chr1_KI270706v1_random': 175055, 'chr1_KI270707v1_random': 32032, 'chr1_KI270708v1_random': 127682, 'chr1_KI270709v1_random': 66860, 'chr1_KI270710v1_random': 40176, 'chr1_KI270711v1_random': 42210, 'chr1_KI270712v1_random': 176043, 'chr1_KI270713v1_random': 40745, 'chr1_KI270714v1_random': 41717, 'chr2_KI270715v1_random': 161471, 'chr2_KI270716v1_random': 153799, 'chr3_GL000221v1_random': 155397, 'chr4_GL000008v2_random': 209709, 'chr5_GL000208v1_random': 92689, 'chr9_KI270717v1_random': 40062, 'chr9_KI270718v1_random': 38054, 'chr9_KI270719v1_random': 176845, 'chr9_KI270720v1_random': 39050, 'chr11_KI270721v1_random': 100316, 'chr14_GL000009v2_random': 201709, 'chr14_GL000225v1_random': 211173, 'chr14_KI270722v1_random': 194050, 'chr14_GL000194v1_random': 191469, 'chr14_KI270723v1_random': 38115, 'chr14_KI270724v1_random': 39555, 'chr14_KI270725v1_random': 172810, 'chr14_KI270726v1_random': 43739, 'chr15_KI270727v1_random': 448248, 'chr16_KI270728v1_random': 1872759, 'chr17_GL000205v2_random': 185591, 'chr17_KI270729v1_random': 280839, 'chr17_KI270730v1_random': 112551, 'chr22_KI270731v1_random': 150754, 'chr22_KI270732v1_random': 41543, 'chr22_KI270733v1_random': 179772, 'chr22_KI270734v1_random': 165050, 'chr22_KI270735v1_random': 42811, 'chr22_KI270736v1_random': 181920, 'chr22_KI270737v1_random': 103838, 'chr22_KI270738v1_random': 99375, 'chr22_KI270739v1_random': 73985, 'chrY_KI270740v1_random': 37240, 'chrUn_KI270302v1': 2274, 'chrUn_KI270304v1': 2165, 'chrUn_KI270303v1': 1942, 'chrUn_KI270305v1': 1472, 'chrUn_KI270322v1': 21476, 'chrUn_KI270320v1': 4416, 'chrUn_KI270310v1': 1201, 'chrUn_KI270316v1': 1444, 'chrUn_KI270315v1': 2276, 'chrUn_KI270312v1': 998, 'chrUn_KI270311v1': 12399, 'chrUn_KI270317v1': 37690, 'chrUn_KI270412v1': 1179, 'chrUn_KI270411v1': 2646, 'chrUn_KI270414v1': 2489, 'chrUn_KI270419v1': 1029, 'chrUn_KI270418v1': 2145, 'chrUn_KI270420v1': 2321, 'chrUn_KI270424v1': 2140, 'chrUn_KI270417v1': 2043, 'chrUn_KI270422v1': 1445, 'chrUn_KI270423v1': 981, 'chrUn_KI270425v1': 1884, 'chrUn_KI270429v1': 1361, 'chrUn_KI270442v1': 392061, 'chrUn_KI270466v1': 1233, 'chrUn_KI270465v1': 1774, 'chrUn_KI270467v1': 3920, 'chrUn_KI270435v1': 92983, 'chrUn_KI270438v1': 112505, 'chrUn_KI270468v1': 4055, 'chrUn_KI270510v1': 2415, 'chrUn_KI270509v1': 2318, 'chrUn_KI270518v1': 2186, 'chrUn_KI270508v1': 1951, 'chrUn_KI270516v1': 1300, 'chrUn_KI270512v1': 22689, 'chrUn_KI270519v1': 138126, 'chrUn_KI270522v1': 5674, 'chrUn_KI270511v1': 8127, 'chrUn_KI270515v1': 6361, 'chrUn_KI270507v1': 5353, 'chrUn_KI270517v1': 3253, 'chrUn_KI270529v1': 1899, 'chrUn_KI270528v1': 2983, 'chrUn_KI270530v1': 2168, 'chrUn_KI270539v1': 993, 'chrUn_KI270538v1': 91309, 'chrUn_KI270544v1': 1202, 'chrUn_KI270548v1': 1599, 'chrUn_KI270583v1': 1400, 'chrUn_KI270587v1': 2969, 'chrUn_KI270580v1': 1553, 'chrUn_KI270581v1': 7046, 'chrUn_KI270579v1': 31033, 'chrUn_KI270589v1': 44474, 'chrUn_KI270590v1': 4685, 'chrUn_KI270584v1': 4513, 'chrUn_KI270582v1': 6504, 'chrUn_KI270588v1': 6158, 'chrUn_KI270593v1': 3041, 'chrUn_KI270591v1': 5796, 'chrUn_KI270330v1': 1652, 'chrUn_KI270329v1': 1040, 'chrUn_KI270334v1': 1368, 'chrUn_KI270333v1': 2699, 'chrUn_KI270335v1': 1048, 'chrUn_KI270338v1': 1428, 'chrUn_KI270340v1': 1428, 'chrUn_KI270336v1': 1026, 'chrUn_KI270337v1': 1121, 'chrUn_KI270363v1': 1803, 'chrUn_KI270364v1': 2855, 'chrUn_KI270362v1': 3530, 'chrUn_KI270366v1': 8320, 'chrUn_KI270378v1': 1048, 'chrUn_KI270379v1': 1045, 'chrUn_KI270389v1': 1298, 'chrUn_KI270390v1': 2387, 'chrUn_KI270387v1': 1537, 'chrUn_KI270395v1': 1143, 'chrUn_KI270396v1': 1880, 'chrUn_KI270388v1': 1216, 'chrUn_KI270394v1': 970, 'chrUn_KI270386v1': 1788, 'chrUn_KI270391v1': 1484, 'chrUn_KI270383v1': 1750, 'chrUn_KI270393v1': 1308, 'chrUn_KI270384v1': 1658, 'chrUn_KI270392v1': 971, 'chrUn_KI270381v1': 1930, 'chrUn_KI270385v1': 990, 'chrUn_KI270382v1': 4215, 'chrUn_KI270376v1': 1136, 'chrUn_KI270374v1': 2656, 'chrUn_KI270372v1': 1650, 'chrUn_KI270373v1': 1451, 'chrUn_KI270375v1': 2378, 'chrUn_KI270371v1': 2805, 'chrUn_KI270448v1': 7992, 'chrUn_KI270521v1': 7642, 'chrUn_GL000195v1': 182896, 'chrUn_GL000219v1': 179198, 'chrUn_GL000220v1': 161802, 'chrUn_GL000224v1': 179693, 'chrUn_KI270741v1': 157432, 'chrUn_GL000226v1': 15008, 'chrUn_GL000213v1': 164239, 'chrUn_KI270743v1': 210658, 'chrUn_KI270744v1': 168472, 'chrUn_KI270745v1': 41891, 'chrUn_KI270746v1': 66486, 'chrUn_KI270747v1': 198735, 'chrUn_KI270748v1': 93321, 'chrUn_KI270749v1': 158759, 'chrUn_KI270750v1': 148850, 'chrUn_KI270751v1': 150742, 'chrUn_KI270752v1': 27745, 'chrUn_KI270753v1': 62944, 'chrUn_KI270754v1': 40191, 'chrUn_KI270755v1': 36723, 'chrUn_KI270756v1': 79590, 'chrUn_KI270757v1': 71251, 'chrUn_GL000214v1': 137718, 'chrUn_KI270742v1': 186739, 'chrUn_GL000216v2': 176608, 'chrUn_GL000218v1': 161147, 'chrEBV': 171823}\n",
      "\tContig chr17\n",
      "16 total jobs\n",
      "chr17:000, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:001, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:002, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:003, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:004, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:005, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:006, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:007, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 26A20G52\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "is_proper_pair True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "VH01429:22:AACFJ5NHV:1:2211:40480:18323\t99\t#16\t43044305\t255\t1S100M\t#16\t43044312\t107\tGTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\tarray('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\t[('MD', '26A20G52'), ('PG', 'MarkDuplicates'), ('RG', 'ET_STAMPExpt_1-CTRL_S2'), ('NH', 1), ('HI', 1), ('NM', 2), ('AS', 190)]\n",
      "reverse_or_forward: -\n",
      "----------------------------\n",
      "MD tag: 26A20G52\n",
      "CIGAR string 1S100M\n",
      "Reference seq: TGCTACCAAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Aligned seq: GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities: array('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities before clipping:\n",
      " array('B', [12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "Soft clipping quality scores ...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Qualities after clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 26, 12, 34, 34, 34, 34, 34, 26, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 26, 34, 34, 26, 34, 34, 34, 34, 26, 34, 34, 34, 34])\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['26', '20', '52']\n",
      "[26, 47, 100]\n",
      "Indicated reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "Fixed aligned seq:\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Finalized fixed aligned seq:\n",
      " tgctaccaagtttatttgcagtgttaCcagcacaacatttacaaaacAtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "Indicated qualities:\n",
      " 34343434343434343434263434343434343434342634343434261234343434342626343434343434343434341234341234343434343434123434343434343412343434343412343434343434343434342634342634343426343426343434342634343434\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['C', 'A']\n",
      "ref bases ['A', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['C', 'A'], ref bases: ['A', 'G']\n",
      "Getting info: C A 12 27\n",
      "Getting info: A G 12 48\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 40G3T55\n",
      "CIGAR tag 100M1S\n",
      "is_reverse True\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "is_proper_pair True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "VH01429:22:AACFJ5NHV:1:2211:40480:18323\t147\t#16\t43044312\t255\t100M1S\t#16\t43044305\t-107\tAAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\tarray('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\t[('MD', '40G3T55'), ('PG', 'MarkDuplicates'), ('RG', 'ET_STAMPExpt_1-CTRL_S2'), ('NH', 1), ('HI', 1), ('NM', 2), ('AS', 190)]\n",
      "reverse_or_forward: -\n",
      "----------------------------\n",
      "MD tag: 40G3T55\n",
      "CIGAR string 100M1S\n",
      "Reference seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Aligned seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "Qualities: array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 100), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "Qualities before clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 12])\n",
      "Soft clipping quality scores ...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Qualities after clipping:\n",
      " array('B', [34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 12, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 26, 34, 34, 34, 26, 34, 34, 34, 34, 34, 34, 34, 34])\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['40', '3', '55']\n",
      "[40, 44, 100]\n",
      "Indicated reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "Fixed aligned seq:\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Finalized fixed aligned seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacAtatGttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "Indicated qualities:\n",
      " 34343434343434263434342634343434341234343434263434343434343434343434343426343434343434341234343434343434343434343434263412343434343434343434342634343434343434343434343434343426343434263434343434343434\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['A', 'G']\n",
      "ref bases ['G', 'T']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['A', 'G'], ref bases: ['G', 'T']\n",
      "chr17:009, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:010, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:011, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:012, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:013, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "chr17:014, total reads: 0, counts_df: Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n",
      "Reads processed:\t2\n",
      "Time to process reads in min:\t0.10064\n",
      "Read Summary:\n",
      "edited         2\n",
      "total_edits    4\n",
      "dtype: int64\n",
      "\n",
      "Deleting overall_label_to_list_of_contents...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit chr17_000_0_5203591, 0/16...\n",
      "\tsplit chr17_010_52035910_57239501, 10/16...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_000_0_5203591', 'chr17_001_5203591_10407182', 'chr17_002_10407182_15610773', 'chr17_003_15610773_20814364', 'chr17_004_20814364_26017955', 'chr17_005_26017955_31221546', 'chr17_006_31221546_36425137', 'chr17_007_36425137_41628728', 'chr17_008_41628728_46832319', 'chr17_009_46832319_52035910', 'chr17_010_52035910_57239501', 'chr17_011_57239501_62443092', 'chr17_012_62443092_67646683', 'chr17_013_67646683_72850274', 'chr17_014_72850274_78053865', 'chr17_015_78053865_83257456'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:05<00:00,  2.73it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!SINGLE END!!!!!\n",
      "~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:05<00:00,  2.78it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    4\n",
      "filtered    3\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.099 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t3\n",
      "\tNumber of unique edit sites:\n",
      "\t3\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "3 sites being converted to SAILOR format...\n",
      "Generating SAILOR-style bed outputs for conversion CT...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "1 final deduplicated C>T SAILOR-formatted sites\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Current memory usage 21.285468MB; Peak: 22.398141MB\n",
      "Time elapsed: 12.79s\n",
      "-------------------------------\n",
      "Deleting intermediate files...\n",
      "-------------------------------\n",
      "++++++\n",
      "Done!\n",
      "++++++\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "tests_folder=\"tests/strandedness_tests/\"\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "for t in \"F1R2_pair_test-single_end_mode_sailor\" \"F1R2_pair_test-single_end_mode\" #\"unstranded_pair_test\" \"F1R2_pair_test\" \"F2R1_end_second_in_pair_test\" \"same_pos_dif_reads_test\" \"tax1bp3_chr17_3665556_read_test\" \"pair_test\"\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running tests...\"\n",
    "    bash tests/strandedness_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "57192b28-b498-4f81-bdcd-0d4d14f644ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "only_5_cells_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/bams/9_3000526_only_5_cells.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/only_5_cells_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tCB\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tFalse\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\t9\n",
      "\tNumber of intervals:\t16\n",
      "\tCores:\t16\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/only_5_cells_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "i_options range(0, 2)\n",
      "j_options range(0, 3)\n",
      "j_options range(3, 6)\n",
      "broken_up_contigs [['9']]\n",
      "Each contig is being split into 16 subsets...\n",
      "\tContig 9\n",
      "16 total jobs\n",
      "Reads processed:\t31\n",
      "Time to process reads in min:\t0.0404\n",
      "Read Summary:\n",
      "edited         31\n",
      "total_edits    99\n",
      "dtype: int64\n",
      "\n",
      "Contigs processed\n",
      "\n",
      "\t['9']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig 9 has 0/4 bams generated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:02<00:00,  6.95it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to concat and write bams: 0.041 minutes\n",
      "Deleting overall_label_to_list_of_contents...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit 9_000_0_7787195, 0/16...\n",
      "\tsplit 9_010_77871950_85659145, 10/16...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:02<00:00,  2.39s/it]\n",
      "100%|██████████| 4/4 [00:02<00:00,  1.74it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    99\n",
      "filtered    99\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.04 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t99\n",
      "\tNumber of unique edit sites:\n",
      "\t73\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed...\n",
      "Current memory usage 21.626729MB; Peak: 22.733417MB\n",
      "Time elapsed: 8.33s\n",
      "-------------------------------\n",
      "Deleting intermediate files...\n",
      "-------------------------------\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "long_read_sc_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 16 cores available for multiprocessing. Set this to the number of available cores for optimal execution.\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//examples/data/LR_single_cell.md.subset.filtered.sorted.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/long_read_sc_test\n",
      "\tBarcode whitelist:\t/tscc/projects/ps-yeolab3/ekofman/sailor2//examples/data/sc_lr_barcodes.tsv.gz\n",
      "\tStrandedness:\t2\n",
      "\tBarcode Tag:\tIB\n",
      "\tPaired End:\tFalse\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tFalse\n",
      "\tBedgraphs:\t[]\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum read quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tMaximum edits per read:\tNone\n",
      "\tContigs:\t6\n",
      "\tNumber of intervals:\t4\n",
      "\tCores:\t16\n",
      "\tVerbose:\tFalse\n",
      "\tKeep intermediate files:\tFalse\n",
      "\tSkip coverage?:\tFalse\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "WARNING /tscc/projects/ps-yeolab3/ekofman/sailor2//tests/singlecell_tests/long_read_sc_test is not empty\n",
      "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "Barcodes in whitelist: 3\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "i_options range(0, 2)\n",
      "j_options range(0, 3)\n",
      "j_options range(3, 6)\n",
      "broken_up_contigs [['6']]\n",
      "Each contig is being split into 4 subsets...\n",
      "\tContig 6\n",
      "4 total jobs\n",
      "Reads processed:\t13\n",
      "Time to process reads in min:\t0.03921\n",
      "Read Summary:\n",
      "edited          3\n",
      "no_edits       10\n",
      "total_edits     3\n",
      "dtype: int64\n",
      "\n",
      "Contigs processed\n",
      "\n",
      "\t['6']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig 6 has 0/4 bams generated\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4/4 [00:02<00:00,  1.78it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t400 suffixes\n",
      "Total time to concat and write bams: 0.282 minutes\n",
      "Deleting overall_label_to_list_of_contents...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites, minimum read quality is 0...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\tsplit 6_000_0_37434137, 0/4...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:16<00:00, 16.82s/it]\n",
      "100%|██████████| 400/400 [00:04<00:00, 80.74it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenating results...\n",
      "Done concatenating.\n",
      "original    3\n",
      "filtered    3\n",
      "dtype: int64\n",
      "Total time to calculate coverage: 0.085 minutes\n",
      "Filtering..\n",
      "\tNumber of edits after filtering:\n",
      "\t3\n",
      "\tNumber of unique edit sites:\n",
      "\t3\n",
      "Writing sites...\n",
      "\n",
      "Adding strand-specific conversion...\n",
      "\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2//annotations/cellranger-mm10-3.0.0.annotation.genes.bed...\n",
      "Current memory usage 21.518014MB; Peak: 22.629871MB\n",
      "Time elapsed: 26.42s\n",
      "-------------------------------\n",
      "Deleting intermediate files...\n",
      "-------------------------------\n",
      "++++++\n",
      "Done!\n",
      "++++++\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "export MARINE=/tscc/projects/ps-yeolab3/ekofman/sailor2/\n",
    "export mypython=/tscc/nfs/home/ekofman/miniconda3/envs/marine_environment/bin/python\n",
    "\n",
    "tests_folder=\"tests/singlecell_tests/\"\n",
    "for t in \"only_5_cells_test\" \"long_read_sc_test\"\n",
    "\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running old tests...\"\n",
    "    bash tests/singlecell_tests/scripts/$t.sh $mypython\n",
    "   \n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8877b85c-bb47-45e7-9ad1-05667d0fbf53",
   "metadata": {},
   "source": [
    "# Integration test automatic checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5c3ed3e-13dd-4399-924e-3d1ac17ce387",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for unstranded_pair_test\n",
      "\tExpecting: {'contig': 'Citrine.dna', 'position': 435, 'count': 22, 'coverage': 22, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '+'}\n",
      "\n",
      "\t >>> unstranded_pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for pair_test\n",
      "\tExpecting: {'contig': '18', 'position': 49491556, 'count': 2, 'coverage': 2, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'RPL17-C18orf32,RPL17', 'feature_strand': '-,-'}\n",
      "\n",
      "\t >>> pair_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '18', 'position': 49567494, 'count': 2, 'coverage': 2, 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '+', 'feature_name': 'LIPG', 'feature_strand': '+'}\n",
      "\n",
      "\t >>> pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'BRCA1', 'feature_strand': '-'}\n",
      "\n",
      "\t >>> F1R2_pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test-single_end_mode_sailor\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 2, 'conversion': 'G>A', 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'BRCA1', 'feature_strand': '-'}\n",
      "\n",
      "\t >>> F1R2_pair_test-single_end_mode_sailor passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test-single_end_mode\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 2, 'conversion': 'G>A', 'num_rows': 1, 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'BRCA1', 'feature_strand': '-'}\n",
      "\n",
      "\t >>> F1R2_pair_test-single_end_mode passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F2R1_end_second_in_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43001716, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'strand_conversion': 'G>A', 'strand': '+', 'feature_name': 'RPL27'}\n",
      "\n",
      "\t >>> F2R1_end_second_in_pair_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for same_pos_dif_reads_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 83199872, 'count': 9, 'coverage': 9, 'conversion': 'C>G', 'strand_conversion': 'C>G', 'strand': '+', 'feature_name': 'AC139099.2'}\n",
      "\n",
      "\t >>> same_pos_dif_reads_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for tax1bp3_chr17_3665556_read_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 3665556, 'num_rows': 1, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'strand_conversion': 'G>A', 'strand': '+'}\n",
      "\n",
      "\t >>> tax1bp3_chr17_3665556_read_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for only_5_cells_test\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GGGACCTTCGAGCCAC-1', 'position': 3000524, 'num_rows': 1, 'count': 1, 'coverage': 12, 'conversion': 'C>A', 'strand_conversion': 'G>T', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GGGACCTTCGAGCCAC-1', 'position': 3000525, 'num_rows': 1, 'count': 1, 'coverage': 12, 'conversion': 'C>T', 'strand_conversion': 'G>A', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GATCCCTCAGTAACGG-1', 'position': 3000525, 'num_rows': 1, 'count': 1, 'coverage': 4, 'conversion': 'C>G', 'strand_conversion': 'G>C', 'strand': '-'}\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for long_read_sc_test\n",
      "\tExpecting: {'contig': '6', 'barcode': 'ENSMUST00000203816-AACGTGTTGGAGAGGG-16-G', 'position': 115807969, 'num_rows': 1, 'count': 1, 'coverage': 1, 'conversion': 'A>C', 'strand_conversion': 'T>G', 'strand': '-', 'feature_name': 'Rpl32'}\n",
      "\n",
      "\t >>> long_read_sc_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '6', 'barcode': 'ENSMUST00000081840-AAGTCGTACCAGGCTC-40-C', 'position': 115805653, 'num_rows': 1, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'strand_conversion': 'C>T', 'strand': '-', 'feature_name': 'Rpl32'}\n",
      "\n",
      "\t >>> long_read_sc_test passed! <<<\n",
      "\n",
      "\tExpecting: {'contig': '6', 'barcode': 'ENSMUST00000081840-AACGTGTTGGAGAGGG-40-G', 'position': 115807015, 'num_rows': 1, 'count': 1, 'coverage': 8, 'conversion': 'C>T', 'strand_conversion': 'G>A', 'strand': '-', 'feature_name': 'Rpl32'}\n",
      "\n",
      "\t >>> long_read_sc_test passed! <<<\n",
      "\n",
      "There were 0 failures\n"
     ]
    }
   ],
   "source": [
    "test_name_to_expectations = {\n",
    "    \"unstranded_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"Citrine.dna\",\n",
    "            \"position\": 435,\n",
    "            \"count\": 22,\n",
    "            \"coverage\": 22,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \"pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49491556,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"RPL17-C18orf32,RPL17\",\n",
    "            \"feature_strand\": \"-,-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"18\",\n",
    "            \"position\": 49567494,\n",
    "            \"count\": 2,\n",
    "            \"coverage\": 2,\n",
    "            \"num_rows\": 1,\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"LIPG\",\n",
    "            \"feature_strand\": \"+\"\n",
    "        }]\n",
    "    },\n",
    "    \n",
    "    \"F1R2_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "    \"F1R2_pair_test-single_end_mode_sailor\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 2,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "    \n",
    "    \"F1R2_pair_test-single_end_mode\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 2,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"num_rows\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\",\n",
    "            \"feature_strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "     \"F2R1_end_second_in_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43001716,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"RPL27\"\n",
    "        }]\n",
    "    },\n",
    "    \"same_pos_dif_reads_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 83199872,\n",
    "            \"count\": 9,\n",
    "            \"coverage\": 9,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"C>G\",\n",
    "            \"strand\": \"+\",\n",
    "            \"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "    \"tax1bp3_chr17_3665556_read_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 3665556,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"+\",\n",
    "            #\"feature_name\": \"AC139099.2\"\n",
    "        }]\n",
    "    },\n",
    "\n",
    "        \"only_5_cells_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000524,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>A\",\n",
    "            \"strand_conversion\": \"G>T\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GATCCCTCAGTAACGG-1\",\n",
    "            \"position\": 3000525,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 4,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"strand_conversion\": \"G>C\",\n",
    "            \"strand\": \"-\"\n",
    "        }]\n",
    "    },\n",
    "    \"long_read_sc_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": [{\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000203816-AACGTGTTGGAGAGGG-16-G\",\n",
    "            \"position\": 115807969,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"A>C\",\n",
    "            \"strand_conversion\": \"T>G\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AAGTCGTACCAGGCTC-40-C\",\n",
    "            \"position\": 115805653,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"strand_conversion\": \"C>T\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        },\n",
    "        {\n",
    "            \"contig\": \"6\",\n",
    "            \"barcode\": \t\"ENSMUST00000081840-AACGTGTTGGAGAGGG-40-G\",\n",
    "            \"position\": 115807015,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 8,\n",
    "            \"conversion\": \"C>T\",\n",
    "            \"strand_conversion\": \"G>A\",\n",
    "            \"strand\": \"-\",\n",
    "            \"feature_name\": \"Rpl32\"\n",
    "        }]\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "failures = 0\n",
    "for test_name, info in test_name_to_expectations.items():\n",
    "    print(\"\\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\\nChecking results for {}\".format(test_name))\n",
    "    \n",
    "    expectations_list = info.get(\"expectations\")\n",
    "    for expectations in expectations_list:\n",
    "        print(\"\\tExpecting: {}\".format(expectations))\n",
    "    \n",
    "              \n",
    "        folder = info.get(\"folder\")\n",
    "        \n",
    "        contig = expectations.get(\"contig\")\n",
    "        barcode = expectations.get(\"barcode\", None)\n",
    "        \n",
    "        position = expectations.get(\"position\")\n",
    "        \n",
    "        final_filtered_site_info_annotated = \"{}/{}/final_filtered_site_info_annotated.tsv\".format(folder, test_name)\n",
    "        final_filtered_site_info_annotated_df = pd.read_csv(final_filtered_site_info_annotated, sep='\\t', index_col=0)\n",
    "    \n",
    "        row_of_interest = final_filtered_site_info_annotated_df[\n",
    "            (final_filtered_site_info_annotated_df['position'] == position) &\\\n",
    "            (final_filtered_site_info_annotated_df['contig'].astype(str) == contig)\n",
    "        ]\n",
    "    \n",
    "    \n",
    "        if barcode:\n",
    "            row_of_interest = row_of_interest[row_of_interest.barcode == barcode]\n",
    "    \n",
    "        failure = False\n",
    "        try:\n",
    "            assert(len(row_of_interest) == expectations.get(\"num_rows\", 1))\n",
    "        except Exception as e:\n",
    "            print(\"Num rows expected: {}, was {}\".format(expectations.get(\"num_rows\", 1), len(row_of_interest)))\n",
    "            failure = True\n",
    "            \n",
    "        for attribute in list(expectations.keys()):\n",
    "            if attribute in ['count', 'coverage', 'conversion', 'strand', 'feature_name']:\n",
    "                attribute_expectation = expectations.get(attribute)\n",
    "                try:\n",
    "                    assert(row_of_interest[attribute].iloc[0] == attribute_expectation)\n",
    "                except Exception as e:\n",
    "                    print(\"Exception: {} was {}\".format(attribute, row_of_interest[attribute].iloc[0]))\n",
    "                    failure = True\n",
    "        if not failure:\n",
    "            print(\"\\n\\t >>> {} passed! <<<\\n\".format(test_name))\n",
    "        else:\n",
    "            print(\"\\n\\t ~~~ {} FAILED! ~~~\\n\".format(test_name))\n",
    "            failures += 1\n",
    "\n",
    "print(\"There were {} failures\".format(failures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b07b328-09c2-4d4e-91c5-f3f22b041648",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fa0520-9e55-4715-ae44-bce13811802a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0bf74f0-f678-4528-ab8e-4bb96e701665",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d014edd-3e67-4b8f-8b97-371a6f7a8ebe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "marine_environment",
   "language": "python",
   "name": "marine_environment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
