{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f2684b1-dbad-45d5-bdb5-b83989bec7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "ba12c2b8-2ccb-4866-86c8-a2284fd1229f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/tests\n",
      "Running\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2\n",
      "F1R2_pair_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 1 cores available for multiprocessing\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/bams/F1R2_pair.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F1R2_pair_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tReverse Stranded:\tFalse\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tContigs:\tchr17\n",
      "\tCores:\t1\n",
      "\tVerbose:\tTrue\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 32 subsets...\n",
      "\tContig chr17\n",
      "32 total jobs\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 26A20G52\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "----------------------------\n",
      "MD tag: 26A20G52\n",
      "CIGAR string 1S100M\n",
      "Reference seq: TGCTACCAAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Aligned seq: GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GTGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['26', '20', '52']\n",
      "[26, 47, 100]\n",
      "Indicated reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tgctaccaagtttatttgcagtgttaAcagcacaacatttacaaaacGtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "Fixed aligned seq:\n",
      " TGCTACCAAGTTTATTTGCAGTGTTACCAGCACAACATTTACAAAACATATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGA\n",
      "Finalized fixed aligned seq:\n",
      " tgctaccaagtttatttgcagtgttaCcagcacaacatttacaaaacAtattttgtacaatcaagtcttcactgcccttgcacactgggggggctaggga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['C', 'A']\n",
      "ref bases ['A', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['C', 'A'], ref bases: ['A', 'G']\n",
      "Getting info: C A 26 27\n",
      "Getting info: A G 34 48\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 40G3T55\n",
      "CIGAR tag 100M1S\n",
      "is_reverse True\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:2211:40480:18323\n",
      "----------------------------\n",
      "MD tag: 40G3T55\n",
      "CIGAR string 100M1S\n",
      "Reference seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACGTATTTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Aligned seq: AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 100), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTAC\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['40', '3', '55']\n",
      "[40, 44, 100]\n",
      "Indicated reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacGtatTttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "Fixed aligned seq:\n",
      " AAGTTTATTTGCAGTGTTAACAGCACAACATTTACAAAACATATGTTGTACAATCAAGTCTTCACTGCCCTTGCACACTGGGGGGGCTAGGGAAGACCTA\n",
      "Finalized fixed aligned seq:\n",
      " aagtttatttgcagtgttaacagcacaacatttacaaaacAtatGttgtacaatcaagtcttcactgcccttgcacactgggggggctagggaagaccta\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['A', 'G']\n",
      "ref bases ['G', 'T']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['A', 'G'], ref bases: ['G', 'T']\n",
      "Getting info: A G 34 41\n",
      "Getting info: G T 12 45\n",
      "\tLength after transpose: 1\n",
      "\tHeight after transpose: 1\n",
      "Reads processed:\t2\n",
      "Time to process reads in min:\t0.0204\n",
      "\n",
      "Contigs processed:\n",
      "\n",
      "\t['chr17']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig chr17 has 0/32 bams generated\n",
      "Will split and reconfigure the following contigs: chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 26.73it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tchr17: num subcontigs to concat: 1\n",
      "\tchr17: concatting\n",
      "\tcontig: chr17 suffix: 0, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:0... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 1, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:1... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 2, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:2... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 3, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:3... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 4, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:4... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 5, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:5... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 6, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:6... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 7, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:7... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 8, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:8... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 9, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:9... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 10, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:10... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 11, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:11... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 12, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:12... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 13, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:13... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 14, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:14... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 15, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:15... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 16, all_contents_df length: 1, all_contents_for_suffix length: 1\n",
      "file chr17_16.bam: 0/2 reads\n",
      "\tcontig: chr17 suffix: 17, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:17... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 18, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:18... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 19, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:19... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 20, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:20... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 21, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:21... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 22, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:22... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 23, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:23... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 24, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:24... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 25, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:25... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 26, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:26... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 27, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:27... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 28, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:28... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 29, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:29... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 30, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:30... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 31, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:31... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "Total time to concat and write bams: 0.029 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\t0/32...\n",
      "\t10/32...\n",
      "\t20/32...\n",
      "\t30/32...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_0', 'chr17_1', 'chr17_2', 'chr17_3', 'chr17_4', 'chr17_5', 'chr17_6', 'chr17_7', 'chr17_8', 'chr17_9', 'chr17_10', 'chr17_11', 'chr17_12', 'chr17_13', 'chr17_14', 'chr17_15', 'chr17_16', 'chr17_17', 'chr17_18', 'chr17_19', 'chr17_20', 'chr17_21', 'chr17_22', 'chr17_23', 'chr17_24', 'chr17_25', 'chr17_26', 'chr17_27', 'chr17_28', 'chr17_29', 'chr17_30', 'chr17_31'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.75s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F1R2_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43044352~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F1R2_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43044331~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F1R2_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43044356~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 27.86it/s]\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/marine.py:159: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  final_site_level_information_df['combo'] = final_site_level_information_df['count'].astype(str) + ',' + final_site_level_information_df['coverage'].astype(str)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to calculate coverage: 0.02 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering and calculating site-level statistics\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Minimum distance from end = 0, Minimum base-calling quality = 0\n",
      "Deduplicating....\n",
      "\tNumber of edits before filtering:\n",
      "\t4\n",
      "\tNumber of edits after filtering:\n",
      "\t3\n",
      "Filtering..\n",
      "\tNumber of unique edit sites:\n",
      "\t3\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Adjusting conversion to feature strand...\n",
      "3 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "F2R1_end_second_in_pair_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 1 cores available for multiprocessing\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/bams/F2R1_end_second_in_pair.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F2R1_end_second_in_pair_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tReverse Stranded:\tFalse\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tContigs:\tchr17\n",
      "\tCores:\t1\n",
      "\tVerbose:\tTrue\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 32 subsets...\n",
      "\tContig chr17\n",
      "32 total jobs\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 10A0G0A84\n",
      "CIGAR tag 13M198556N84M1S\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:1112:41030:25422\n",
      "----------------------------\n",
      "MD tag: 10A0G0A84\n",
      "CIGAR string 13M198556N84M1S\n",
      "Reference seq: GGCTGTCATGAGAGAGGAGGAGGATGAGCTCAAAGATGAAGTTCAAAGTCAGTCCTCTGCTTCCTCAGAGGATTACATCATCATCCTGCCTGAGTGC\n",
      "Aligned seq: GGCTGTCATGGAGGAGGAGGAGGATGAGCTCAAAGATGAAGTTCAAAGTCAGTCCTCTGCTTCCTCAGAGGATTACATCATCATCCTGCCTGAGTGCA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 13), (3, 198556), (0, 84), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GGCTGTCATGGAGGAGGAGGAGGATGAGCTCAAAGATGAAGTTCAAAGTCAGTCCTCTGCTTCCTCAGAGGATTACATCATCATCCTGCCTGAGTGCA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 13), (3, 198556), (0, 84)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " GGCTGTCATGGAGGAGGAGGAGGATGAGCTCAAAGATGAAGTTCAAAGTCAGTCCTCTGCTTCCTCAGAGGATTACATCATCATCCTGCCTGAGTGC\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['10', '0', '0', '84']\n",
      "[10, 11, 12, 97]\n",
      "Indicated reference seq:\n",
      " ggctgtcatgAGAgaggaggaggatgagctcaaagatgaagttcaaagtcagtcctctgcttcctcagaggattacatcatcatcctgcctgagtgc\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ggctgtcatgAGA198556*ngaggaggaggatgagctcaaagatgaagttcaaagtcagtcctctgcttcctcagaggattacatcatcatcctgcctgagtgc\n",
      "Fixed aligned seq:\n",
      " GGCTGTCATGGAG198556*nGAGGAGGAGGATGAGCTCAAAGATGAAGTTCAAAGTCAGTCCTCTGCTTCCTCAGAGGATTACATCATCATCCTGCCTGAGTGC\n",
      "Finalized fixed aligned seq:\n",
      " ggctgtcatgGAG198556*ngaggaggaggatgagctcaaagatgaagttcaaagtcagtcctctgcttcctcagaggattacatcatcatcctgcctgagtgc\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A', 'G']\n",
      "ref bases ['A', 'G', 'A']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A', 'G'], ref bases: ['A', 'G', 'A']\n",
      "Getting info: G A 26 11\n",
      "Getting info: A G 34 12\n",
      "Getting info: G A 34 13\n",
      "\tLength after transpose: 1\n",
      "\tHeight after transpose: 1\n",
      "Reads processed:\t1\n",
      "Time to process reads in min:\t0.0218\n",
      "\n",
      "Contigs processed:\n",
      "\n",
      "\t['chr17']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig chr17 has 0/32 bams generated\n",
      "Will split and reconfigure the following contigs: chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 25.38it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tchr17: num subcontigs to concat: 1\n",
      "\tchr17: concatting\n",
      "\tcontig: chr17 suffix: 0, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:0... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 1, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:1... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 2, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:2... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 3, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:3... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 4, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:4... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 5, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:5... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 6, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:6... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 7, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:7... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 8, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:8... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 9, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:9... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 10, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:10... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 11, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:11... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 12, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:12... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 13, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:13... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 14, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:14... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 15, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:15... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 16, all_contents_df length: 1, all_contents_for_suffix length: 1\n",
      "file chr17_16.bam: 0/1 reads\n",
      "\tcontig: chr17 suffix: 17, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:17... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 18, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:18... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 19, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:19... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 20, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:20... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 21, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:21... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 22, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:22... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 23, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:23... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 24, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:24... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 25, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:25... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 26, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:26... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 27, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:27... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 28, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:28... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 29, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:29... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 30, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:30... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 31, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:31... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "Total time to concat and write bams: 0.029 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\t0/32...\n",
      "\t10/32...\n",
      "\t20/32...\n",
      "\t30/32...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_0', 'chr17_1', 'chr17_2', 'chr17_3', 'chr17_4', 'chr17_5', 'chr17_6', 'chr17_7', 'chr17_8', 'chr17_9', 'chr17_10', 'chr17_11', 'chr17_12', 'chr17_13', 'chr17_14', 'chr17_15', 'chr17_16', 'chr17_17', 'chr17_18', 'chr17_19', 'chr17_20', 'chr17_21', 'chr17_22', 'chr17_23', 'chr17_24', 'chr17_25', 'chr17_26', 'chr17_27', 'chr17_28', 'chr17_29', 'chr17_30', 'chr17_31'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.73s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F2R1_end_second_in_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43001715~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F2R1_end_second_in_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43001716~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/F2R1_end_second_in_pair_test/split_bams/chr17_16/chr17_16.bam.sorted.bam\n",
      " chr17:43001717~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 28.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to calculate coverage: 0.019 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering and calculating site-level statistics\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Minimum distance from end = 0, Minimum base-calling quality = 0\n",
      "Deduplicating....\n",
      "\tNumber of edits before filtering:\n",
      "\t3\n",
      "\tNumber of edits after filtering:\n",
      "\t3\n",
      "Filtering..\n",
      "\tNumber of unique edit sites:\n",
      "\t3\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Adjusting conversion to feature strand...\n",
      "3 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "same_pos_dif_reads_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 1 cores available for multiprocessing\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/bams/same_pos_dif_reads.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tReverse Stranded:\tFalse\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tContigs:\tchr17\n",
      "\tCores:\t1\n",
      "\tVerbose:\tTrue\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 32 subsets...\n",
      "\tContig chr17\n",
      "32 total jobs\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 55C43\n",
      "CIGAR tag 1S97M17144N2M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:2:2606:39477:40189\n",
      "----------------------------\n",
      "MD tag: 55C43\n",
      "CIGAR string 1S97M17144N2M\n",
      "Reference seq: CCCACCTTCGCTTCCGGGACCACAGAGCCCCTGGAGCCTGCACGCCACGGAGTCTCCCACTGAGGAGTGAGGAGGCCTCTTGTGGGGACAGATCTGGGA\n",
      "Aligned seq: TCCCACCTTCGCTTCCGGGACCACAGAGCCCCTGGAGCCTGCACGCCACGGAGTCTACCACTGAGGAGTGAGGAGGCCTCTTGTGGGGACAGATCTGGGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 97), (3, 17144), (0, 2)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TCCCACCTTCGCTTCCGGGACCACAGAGCCCCTGGAGCCTGCACGCCACGGAGTCTACCACTGAGGAGTGAGGAGGCCTCTTGTGGGGACAGATCTGGGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 97), (3, 17144), (0, 2)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CCCACCTTCGCTTCCGGGACCACAGAGCCCCTGGAGCCTGCACGCCACGGAGTCTACCACTGAGGAGTGAGGAGGCCTCTTGTGGGGACAGATCTGGGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['55', '43']\n",
      "[55, 99]\n",
      "Indicated reference seq:\n",
      " cccaccttcgcttccgggaccacagagcccctggagcctgcacgccacggagtctCccactgaggagtgaggaggcctcttgtggggacagatctggga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " cccaccttcgcttccgggaccacagagcccctggagcctgcacgccacggagtctCccactgaggagtgaggaggcctcttgtggggacagatctgg17144*nga\n",
      "Fixed aligned seq:\n",
      " CCCACCTTCGCTTCCGGGACCACAGAGCCCCTGGAGCCTGCACGCCACGGAGTCTACCACTGAGGAGTGAGGAGGCCTCTTGTGGGGACAGATCTGG17144*nGA\n",
      "Finalized fixed aligned seq:\n",
      " cccaccttcgcttccgggaccacagagcccctggagcctgcacgccacggagtctAccactgaggagtgaggaggcctcttgtggggacagatctgg17144*nga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['A']\n",
      "ref bases ['C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['A'], ref bases: ['C']\n",
      "Getting info: A C 34 56\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 86C13\n",
      "CIGAR tag 1S17M4579N83M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:2:1602:67558:13893\n",
      "----------------------------\n",
      "MD tag: 86C13\n",
      "CIGAR string 1S17M4579N83M\n",
      "Reference seq: TGTGGGGACAGATCTGGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAAT\n",
      "Aligned seq: GTGTGGGGACAGATCTGGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAAT\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 17), (3, 4579), (0, 83)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GTGTGGGGACAGATCTGGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAAT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 17), (3, 4579), (0, 83)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TGTGGGGACAGATCTGGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAAT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['86', '13']\n",
      "[86, 100]\n",
      "Indicated reference seq:\n",
      " tgtggggacagatctggaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaat\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tgtggggacagatctgg4579*naatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaat\n",
      "Fixed aligned seq:\n",
      " TGTGGGGACAGATCTGG4579*nAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAAT\n",
      "Finalized fixed aligned seq:\n",
      " tgtggggacagatctgg4579*naatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaat\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G']\n",
      "ref bases ['C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G'], ref bases: ['C']\n",
      "Getting info: G C 34 4666\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 95C4\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:1109:44703:39053\n",
      "----------------------------\n",
      "MD tag: 95C4\n",
      "CIGAR string 1S100M\n",
      "Reference seq: CTATTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTG\n",
      "Aligned seq: GCTATTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTG\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GCTATTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTG\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CTATTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTG\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['95', '4']\n",
      "[95, 100]\n",
      "Indicated reference seq:\n",
      " ctattattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtg\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ctattattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtg\n",
      "Fixed aligned seq:\n",
      " CTATTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTG\n",
      "Finalized fixed aligned seq:\n",
      " ctattattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtg\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G']\n",
      "ref bases ['C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G'], ref bases: ['C']\n",
      "Getting info: G C 34 96\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 92C8\n",
      "CIGAR tag 101M\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:1208:38151:26028\n",
      "----------------------------\n",
      "MD tag: 92C8\n",
      "CIGAR string 101M\n",
      "Reference seq: TTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGA\n",
      "Aligned seq: TTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['92', '8']\n",
      "[92, 101]\n",
      "Indicated reference seq:\n",
      " ttattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ttattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcga\n",
      "Fixed aligned seq:\n",
      " TTATTGAGTTTCTTCTTTTTCAGAATGAGTGGCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGA\n",
      "Finalized fixed aligned seq:\n",
      " ttattgagtttcttctttttcagaatgagtggcatccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G']\n",
      "ref bases ['C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G'], ref bases: ['C']\n",
      "Getting info: G C 34 93\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 60C37\n",
      "CIGAR tag 1S98M2S\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:2112:47146:35399\n",
      "----------------------------\n",
      "MD tag: 60C37\n",
      "CIGAR string 1S98M2S\n",
      "Reference seq: CATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCA\n",
      "Aligned seq: TCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAAT\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 98), (4, 2)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TCATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAAT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 98)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['60', '37']\n",
      "[60, 98]\n",
      "Indicated reference seq:\n",
      " catccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgca\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " catccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgca\n",
      "Fixed aligned seq:\n",
      " CATCCAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCA\n",
      "Finalized fixed aligned seq:\n",
      " catccaaacctgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgca\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G']\n",
      "ref bases ['C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G'], ref bases: ['C']\n",
      "Getting info: G C 34 61\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 5C50C37G5\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:2:2610:69319:26899\n",
      "----------------------------\n",
      "MD tag: 5C50C37G5\n",
      "CIGAR string 1S100M\n",
      "Reference seq: CAAACCTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTT\n",
      "Aligned seq: TCAAACTTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTT\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TCAAACTTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CAAACTTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['5', '50', '37', '5']\n",
      "[5, 56, 94, 100]\n",
      "Indicated reference seq:\n",
      " caaacCtgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactt\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " caaacCtgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactt\n",
      "Fixed aligned seq:\n",
      " CAAACTTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTT\n",
      "Finalized fixed aligned seq:\n",
      " caaacTtgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactt\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['T', 'G', 'A']\n",
      "ref bases ['C', 'C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['T', 'G', 'A'], ref bases: ['C', 'C', 'G']\n",
      "Getting info: T C 34 6\n",
      "Getting info: G C 34 57\n",
      "Getting info: A G 34 95\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 50C37G10\n",
      "CIGAR tag 1S99M1S\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:2610:69319:26899\n",
      "----------------------------\n",
      "MD tag: 50C37G10\n",
      "CIGAR string 1S99M1S\n",
      "Reference seq: TGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTG\n",
      "Aligned seq: TTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 99), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TTGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 99)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTG\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['50', '37', '10']\n",
      "[50, 88, 99]\n",
      "Indicated reference seq:\n",
      " tgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctg\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tgactcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctg\n",
      "Fixed aligned seq:\n",
      " TGACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTG\n",
      "Finalized fixed aligned seq:\n",
      " tgactcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctg\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 51\n",
      "Getting info: A G 34 89\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 48C37G14\n",
      "CIGAR tag 101M\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:1403:49513:32843\n",
      "----------------------------\n",
      "MD tag: 48C37G14\n",
      "CIGAR string 101M\n",
      "Reference seq: ACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAA\n",
      "Aligned seq: ACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " ACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " ACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['48', '37', '14']\n",
      "[48, 86, 101]\n",
      "Indicated reference seq:\n",
      " actcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " actcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaa\n",
      "Fixed aligned seq:\n",
      " ACTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "Finalized fixed aligned seq:\n",
      " actcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 49\n",
      "Getting info: A G 34 87\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 47C37G14\n",
      "CIGAR tag 1S100M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:1:2414:26942:20235\n",
      "----------------------------\n",
      "MD tag: 47C37G14\n",
      "CIGAR string 1S100M\n",
      "Reference seq: CTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAA\n",
      "Aligned seq: TCTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 1), (0, 100)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TCTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['47', '37', '14']\n",
      "[47, 85, 100]\n",
      "Indicated reference seq:\n",
      " ctcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ctcatatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaa\n",
      "Fixed aligned seq:\n",
      " CTCATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAA\n",
      "Finalized fixed aligned seq:\n",
      " ctcatatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 48\n",
      "Getting info: A G 34 86\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 2T9T0T26G4C37G1A11\n",
      "CIGAR tag 3S97M\n",
      "is_reverse False\n",
      "is_read1 False\n",
      "is_read2 True\n",
      "is_paired True\n",
      "mate_is_reverse True\n",
      "read id VH01429:22:AACFJ5NHV:2:1608:28097:17679\n",
      "----------------------------\n",
      "MD tag: 2T9T0T26G4C37G1A11\n",
      "CIGAR string 3S97M\n",
      "Reference seq: CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAA\n",
      "Aligned seq: CCACAAATTTCCAACAATATTAATTTTTGCAAGACTTCGGAGGCTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATTCTTTCCTGTAA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 3), (0, 97)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " CCACAAATTTCCAACAATATTAATTTTTGCAAGACTTCGGAGGCTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATTCTTTCCTGTAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 97)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CAAATTTCCAACAATATTAATTTTTGCAAGACTTCGGAGGCTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATTCTTTCCTGTAA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['2', '9', '0', '26', '4', '37', '1', '11']\n",
      "[2, 12, 13, 40, 45, 83, 85, 97]\n",
      "Indicated reference seq:\n",
      " caTatttccaacTTtattaatttttgcaagacttcggaggGtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtActttcctgtaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " caTatttccaacTTtattaatttttgcaagacttcggaggGtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtActttcctgtaa\n",
      "Fixed aligned seq:\n",
      " CAAATTTCCAACAATATTAATTTTTGCAAGACTTCGGAGGCTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATTCTTTCCTGTAA\n",
      "Finalized fixed aligned seq:\n",
      " caAatttccaacAAtattaatttttgcaagacttcggaggCtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtTctttcctgtaa\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['A', 'A', 'A', 'C', 'G', 'A', 'T']\n",
      "ref bases ['T', 'T', 'T', 'G', 'C', 'G', 'A']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['A', 'A', 'A', 'C', 'G', 'A', 'T'], ref bases: ['T', 'T', 'T', 'G', 'C', 'G', 'A']\n",
      "Getting info: A T 12 3\n",
      "Getting info: A T 34 13\n",
      "Getting info: A T 34 14\n",
      "Getting info: C G 26 41\n",
      "Getting info: G C 34 46\n",
      "Getting info: A G 34 84\n",
      "Getting info: T A 34 86\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 45C37G16\n",
      "CIGAR tag 100M1S\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:1608:28097:17679\n",
      "----------------------------\n",
      "MD tag: 45C37G16\n",
      "CIGAR string 100M1S\n",
      "Reference seq: CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAAAC\n",
      "Aligned seq: CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACG\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 100), (4, 1)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACG\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 100)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAAC\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['45', '37', '16']\n",
      "[45, 83, 100]\n",
      "Indicated reference seq:\n",
      " catatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaac\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " catatttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaac\n",
      "Fixed aligned seq:\n",
      " CATATTTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAAC\n",
      "Finalized fixed aligned seq:\n",
      " catatttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaaac\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 26 46\n",
      "Getting info: A G 26 84\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 40C37G20\n",
      "CIGAR tag 99M2S\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:1104:58090:46985\n",
      "----------------------------\n",
      "MD tag: 40C37G20\n",
      "CIGAR string 99M2S\n",
      "Reference seq: TTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAAACTGAC\n",
      "Aligned seq: TTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACCT\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 99), (4, 2)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACCT\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 99)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGAC\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['40', '37', '20']\n",
      "[40, 78, 99]\n",
      "Indicated reference seq:\n",
      " ttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgac\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ttccaactttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgac\n",
      "Fixed aligned seq:\n",
      " TTCCAACTTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGAC\n",
      "Finalized fixed aligned seq:\n",
      " ttccaactttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaaactgac\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 41\n",
      "Getting info: A G 34 79\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 33C37G19C9\n",
      "CIGAR tag 101M\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:1109:44703:39053\n",
      "----------------------------\n",
      "MD tag: 33C37G19C9\n",
      "CIGAR string 101M\n",
      "Reference seq: TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "Aligned seq: TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGATTAAAGGAGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGATTAAAGGAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGATTAAAGGAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['33', '37', '19', '9']\n",
      "[33, 71, 91, 101]\n",
      "Indicated reference seq:\n",
      " tttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgaCtaaaggaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgaCtaaaggaga\n",
      "Fixed aligned seq:\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGATTAAAGGAGA\n",
      "Finalized fixed aligned seq:\n",
      " tttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaaactgaTtaaaggaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A', 'T']\n",
      "ref bases ['C', 'G', 'C']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A', 'T'], ref bases: ['C', 'G', 'C']\n",
      "Getting info: G C 34 34\n",
      "Getting info: A G 34 72\n",
      "Getting info: T C 34 92\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 33C37G29\n",
      "CIGAR tag 101M\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:1:2414:26942:20235\n",
      "----------------------------\n",
      "MD tag: 33C37G29\n",
      "CIGAR string 101M\n",
      "Reference seq: TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "Aligned seq: TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 101)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['33', '37', '29']\n",
      "[33, 71, 101]\n",
      "Indicated reference seq:\n",
      " tttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgactaaaggaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " tttattaatttttgcaagacttcggagggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgactaaaggaga\n",
      "Fixed aligned seq:\n",
      " TTTATTAATTTTTGCAAGACTTCGGAGGGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGA\n",
      "Finalized fixed aligned seq:\n",
      " tttattaatttttgcaagacttcggagggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaaactgactaaaggaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 34\n",
      "Getting info: A G 34 72\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 6C37G53\n",
      "CIGAR tag 98M2S\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:1602:67558:13893\n",
      "----------------------------\n",
      "MD tag: 6C37G53\n",
      "CIGAR string 98M2S\n",
      "Reference seq: GGTGCTCTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAGTACTTTCCTGTAAAACTGACTAAAGGAGAAGTTGGTGTCTTCCTCATACAAGA\n",
      "Aligned seq: GGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGAAGTTGGTGTCTTCCTCATACAAGAGA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(0, 98), (4, 2)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " GGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGAAGTTGGTGTCTTCCTCATACAAGAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 98)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " GGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGAAGTTGGTGTCTTCCTCATACAAGA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['6', '37', '53']\n",
      "[6, 44, 98]\n",
      "Indicated reference seq:\n",
      " ggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgactaaaggagaagttggtgtcttcctcatacaaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " ggtgctCtgtggcgattaatattgctgagctgagcacagctgcaGtactttcctgtaaaactgactaaaggagaagttggtgtcttcctcatacaaga\n",
      "Fixed aligned seq:\n",
      " GGTGCTGTGTGGCGATTAATATTGCTGAGCTGAGCACAGCTGCAATACTTTCCTGTAAAACTGACTAAAGGAGAAGTTGGTGTCTTCCTCATACAAGA\n",
      "Finalized fixed aligned seq:\n",
      " ggtgctGtgtggcgattaatattgctgagctgagcacagctgcaAtactttcctgtaaaactgactaaaggagaagttggtgtcttcctcatacaaga\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'A']\n",
      "ref bases ['C', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'A'], ref bases: ['C', 'G']\n",
      "Getting info: G C 34 7\n",
      "Getting info: A G 34 45\n",
      "\tLength after transpose: 1\n",
      "\tHeight after transpose: 1\n",
      "Reads processed:\t16\n",
      "Time to process reads in min:\t0.0212\n",
      "\n",
      "Contigs processed:\n",
      "\n",
      "\t['chr17']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig chr17 has 0/32 bams generated\n",
      "Will split and reconfigure the following contigs: chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 25.76it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tchr17: num subcontigs to concat: 1\n",
      "\tchr17: concatting\n",
      "\tcontig: chr17 suffix: 0, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:0... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 1, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:1... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 2, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:2... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 3, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:3... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 4, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:4... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 5, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:5... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 6, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:6... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 7, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:7... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 8, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:8... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 9, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:9... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 10, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:10... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 11, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:11... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 12, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:12... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 13, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:13... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 14, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:14... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 15, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:15... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 16, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:16... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 17, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:17... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 18, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:18... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 19, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:19... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 20, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:20... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 21, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:21... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 22, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:22... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 23, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:23... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 24, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:24... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 25, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:25... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 26, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:26... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 27, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:27... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 28, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:28... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 29, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:29... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 30, all_contents_df length: 1, all_contents_for_suffix length: 0\n",
      "\tWARNING: No reads found in region chr17:30... assuming this is not an issue, but perhaps worth confirming manually using samtools.\n",
      "\tcontig: chr17 suffix: 31, all_contents_df length: 1, all_contents_for_suffix length: 1\n",
      "file chr17_31.bam: 0/16 reads\n",
      "Total time to concat and write bams: 0.033 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\t0/32...\n",
      "\t10/32...\n",
      "\t20/32...\n",
      "\t30/32...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_0', 'chr17_1', 'chr17_2', 'chr17_3', 'chr17_4', 'chr17_5', 'chr17_6', 'chr17_7', 'chr17_8', 'chr17_9', 'chr17_10', 'chr17_11', 'chr17_12', 'chr17_13', 'chr17_14', 'chr17_15', 'chr17_16', 'chr17_17', 'chr17_18', 'chr17_19', 'chr17_20', 'chr17_21', 'chr17_22', 'chr17_23', 'chr17_24', 'chr17_25', 'chr17_26', 'chr17_27', 'chr17_28', 'chr17_29', 'chr17_30', 'chr17_31'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.93s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199840~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199872~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199910~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199912~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199821~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83195182~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199829~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199930~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199867~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/same_pos_dif_reads_test/split_bams/chr17_31/chr17_31.bam.sorted.bam\n",
      " chr17:83199839~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 26.78it/s]\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/marine.py:159: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  final_site_level_information_df['combo'] = final_site_level_information_df['count'].astype(str) + ',' + final_site_level_information_df['coverage'].astype(str)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to calculate coverage: 0.02 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering and calculating site-level statistics\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Minimum distance from end = 0, Minimum base-calling quality = 0\n",
      "Deduplicating....\n",
      "\tNumber of edits before filtering:\n",
      "\t32\n",
      "\tNumber of edits after filtering:\n",
      "\t24\n",
      "Filtering..\n",
      "\tNumber of unique edit sites:\n",
      "\t10\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Adjusting conversion to feature strand...\n",
      "10 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "++++++\n",
      "Done!\n",
      "++++++\n",
      "tax1bp3_chr17_3665556_read_test\n",
      "Removing old files...\n",
      "Running old tests...\n",
      "Assuming 1 cores available for multiprocessing\n",
      "Arguments:\n",
      "\tBAM filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/bams/tax1bp3_chr17_3665556_read.bam\n",
      "\tAnnotation bedfile filepath:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed\n",
      "\tOutput folder:\t/tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test\n",
      "\tBarcode whitelist:\tNone\n",
      "\tReverse Stranded:\tFalse\n",
      "\tBarcode Tag:\tNone\n",
      "\tPaired End:\tTrue\n",
      "\tCoverage only:\tFalse\n",
      "\tFiltering only:\tFalse\n",
      "\tAnnotation only:\tFalse\n",
      "\tSailor outputs:\tTrue\n",
      "\tMinimum base quality:\t0\n",
      "\tMinimum distance from end:\t0\n",
      "\tContigs:\tchr17\n",
      "\tCores:\t1\n",
      "\tVerbose:\tTrue\n",
      "\n",
      "::::    ::::      :::     :::::::::  ::::::::::: ::::    ::: :::::::::: \n",
      "+:+:+: :+:+:+   :+: :+:   :+:    :+:     :+:     :+:+:   :+: :+:        \n",
      "+:+ +:+:+ +:+  +:+   +:+  +:+    +:+     +:+     :+:+:+  +:+ +:+        \n",
      "+#+  +:+  +#+ +#++:++#++: +#++:++#:      +#+     +#+ +:+ +#+ +#++:++#   \n",
      "+#+       +#+ +#+     +#+ +#+    +#+     +#+     +#+  +#+#+# +#+        \n",
      "#+#       #+# #+#     #+# #+#    #+#     #+#     #+#   #+#+# #+#        \n",
      "###       ### ###     ### ###    ### ########### ###    #### ########## \n",
      "==================================================================\n",
      "Multi-core Algorithm for Rapid Identification of Nucleotide Edits\n",
      "==================================================================\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Identifying edits\n",
      "~~~~~~~~~~~~~~~~~~\n",
      "Each contig is being split into 32 subsets...\n",
      "\tContig chr17\n",
      "32 total jobs\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "MD tag 2C1C6G3A1G4G47\n",
      "CIGAR tag 28S70M\n",
      "is_reverse True\n",
      "is_read1 True\n",
      "is_read2 False\n",
      "is_paired True\n",
      "mate_is_reverse False\n",
      "read id VH01429:22:AACFJ5NHV:2:2414:21053:15521\n",
      "----------------------------\n",
      "MD tag: 2C1C6G3A1G4G47\n",
      "CIGAR string 28S70M\n",
      "Reference seq: AACGCGTGAAGGAAAATGATCAGAAAAAGAAAGAAGCCAAAGAGAAAGGTACCTGGGTTCAACTAAAGCA\n",
      "Aligned seq: CACAATAAGAGCCTATTTTGCTTCCTTAAAGGTGTGAAGTAAATTAATCATAAAAAGAAAGAAGCCAAAGAGAAAGGTACCTGGGTTCAACTAAAGCA\n",
      "CIGAR tuples before clipping (if needed):\n",
      " [(4, 28), (0, 70)]\n",
      "Aligned sequence before clipping (if needed):\n",
      " CACAATAAGAGCCTATTTTGCTTCCTTAAAGGTGTGAAGTAAATTAATCATAAAAAGAAAGAAGCCAAAGAGAAAGGTACCTGGGTTCAACTAAAGCA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "CIGAR tuples after clipping (if needed):\n",
      " [(0, 70)]\n",
      "Aligned sequence after clipping (if needed):\n",
      " AAGGTGTGAAGTAAATTAATCATAAAAAGAAAGAAGCCAAAGAGAAAGGTACCTGGGTTCAACTAAAGCA\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "['2', '1', '6', '3', '1', '4', '47']\n",
      "[2, 4, 11, 15, 17, 22, 70]\n",
      "Indicated reference seq:\n",
      " aaCgCgtgaagGaaaAtGatcaGaaaaagaaagaagccaaagagaaaggtacctgggttcaactaaagca\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Fixed reference seq:\n",
      " aaCgCgtgaagGaaaAtGatcaGaaaaagaaagaagccaaagagaaaggtacctgggttcaactaaagca\n",
      "Fixed aligned seq:\n",
      " AAGGTGTGAAGTAAATTAATCATAAAAAGAAAGAAGCCAAAGAGAAAGGTACCTGGGTTCAACTAAAGCA\n",
      "Finalized fixed aligned seq:\n",
      " aaGgTgtgaagTaaaTtAatcaTaaaaagaaagaagccaaagagaaaggtacctgggttcaactaaagca\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "alt bases ['G', 'T', 'T', 'T', 'A', 'T']\n",
      "ref bases ['C', 'C', 'G', 'A', 'G', 'G']\n",
      "Successfully ran get_edit_information_wrapper\n",
      "alt bases: ['G', 'T', 'T', 'T', 'A', 'T'], ref bases: ['C', 'C', 'G', 'A', 'G', 'G']\n",
      "Getting info: G C 34 3\n",
      "Getting info: T C 12 5\n",
      "Getting info: T G 34 12\n",
      "Getting info: T A 26 16\n",
      "Getting info: A G 34 18\n",
      "Getting info: T G 34 23\n",
      "\tLength after transpose: 1\n",
      "\tHeight after transpose: 1\n",
      "Reads processed:\t1\n",
      "Time to process reads in min:\t0.0186\n",
      "\n",
      "Contigs processed:\n",
      "\n",
      "\t['chr17']\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Splitting and reconfiguring BAMs to optimize coverage calculations\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Contig chr17 has 0/32 bams generated\n",
      "Will split and reconfigure the following contigs: chr17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 29.36it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to concat and write bams: 0.03 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Calculating coverage at edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\t0/32...\n",
      "\t10/32...\n",
      "\t20/32...\n",
      "\t30/32...\n",
      "Done grouping! Concatenating ...\n",
      "Done concatenating!\n",
      "edit_info_grouped_per_contig_combined dict_keys(['chr17_0', 'chr17_1', 'chr17_2', 'chr17_3', 'chr17_4', 'chr17_5', 'chr17_6', 'chr17_7', 'chr17_8', 'chr17_9', 'chr17_10', 'chr17_11', 'chr17_12', 'chr17_13', 'chr17_14', 'chr17_15', 'chr17_16', 'chr17_17', 'chr17_18', 'chr17_19', 'chr17_20', 'chr17_21', 'chr17_22', 'chr17_23', 'chr17_24', 'chr17_25', 'chr17_26', 'chr17_27', 'chr17_28', 'chr17_29', 'chr17_30', 'chr17_31'])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.81s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665541~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665543~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665550~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665554~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665556~~~~~~~`\n",
      "~~~~~~\n",
      "!!!!PAIRED END!!!!!\n",
      "pos: /tscc/projects/ps-yeolab3/ekofman/sailor2/tests/strandedness_tests/tax1bp3_chr17_3665556_read_test/split_bams/chr17_1/chr17_1.bam.sorted.bam\n",
      " chr17:3665561~~~~~~~`\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:01<00:00, 29.65it/s]\n",
      "/tscc/projects/ps-yeolab3/ekofman/sailor2/marine.py:159: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  final_site_level_information_df['combo'] = final_site_level_information_df['count'].astype(str) + ',' + final_site_level_information_df['coverage'].astype(str)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time to calculate coverage: 0.018 minutes\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering and calculating site-level statistics\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Filtering edited sites\n",
      "~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Minimum distance from end = 0, Minimum base-calling quality = 0\n",
      "Deduplicating....\n",
      "\tNumber of edits before filtering:\n",
      "\t6\n",
      "\tNumber of edits after filtering:\n",
      "\t6\n",
      "Filtering..\n",
      "\tNumber of unique edit sites:\n",
      "\t6\n",
      "Annotating sites with GTF information from /tscc/projects/ps-yeolab3/ekofman/sailor2/annotations/hg38_gencode.v35.annotation.genes.bed...\n",
      "Adjusting conversion to feature strand...\n",
      "6 sites being converted to SAILOR format...\n",
      "0 rows had coverage of 0 or more edits than coverage... filtering these out, but look into them...\n",
      "++++++\n",
      "Done!\n",
      "++++++\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "tests_folder=\"tests/strandedness_tests/\"\n",
    "for t in \"F1R2_pair_test\" \"F2R1_end_second_in_pair_test\" \"same_pos_dif_reads_test\" \"tax1bp3_chr17_3665556_read_test\"\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running old tests...\"\n",
    "    bash tests/strandedness_tests/scripts/$t.sh \n",
    "   \n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57192b28-b498-4f81-bdcd-0d4d14f644ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "echo \"Running\"\n",
    "\n",
    "cd ..\n",
    "pwd\n",
    "\n",
    "tests_folder=\"tests/singlecell_tests/\"\n",
    "for t in \"only_5_cells_test\"\n",
    "do\n",
    "    echo $t\n",
    "    echo \"Removing old files...\"\n",
    "    rm $tests_folder$t/* -r\n",
    "\n",
    "    echo \"Running old tests...\"\n",
    "    bash tests/singlecell_tests/scripts/$t.sh \n",
    "   \n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8877b85c-bb47-45e7-9ad1-05667d0fbf53",
   "metadata": {},
   "source": [
    "# Integration test automatic checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5c3ed3e-13dd-4399-924e-3d1ac17ce387",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_name_to_expectations = {\n",
    "    \"only_5_cells_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000527,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 10,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"T>C\"\n",
    "        }\n",
    "    },\n",
    "\n",
    "    \"only_5_cells_test\": {\n",
    "        \"folder\": \"singlecell_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"9\",\n",
    "            \"barcode\": \t\"GGGACCTTCGAGCCAC-1\",\n",
    "            \"position\": 3000527,\n",
    "            \"num_rows\": 1,\n",
    "            \"count\": 10,\n",
    "            \"coverage\": 12,\n",
    "            \"conversion\": \"T>C\"\n",
    "        }\n",
    "    },\n",
    "    \n",
    "    \"F1R2_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43044352,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"feature_conversion\": \"C>T\",\n",
    "            \"feature_strand\": \"-\",\n",
    "            \"feature_name\": \"BRCA1\"\n",
    "        }\n",
    "    },\n",
    "\n",
    "     \"F2R1_end_second_in_pair_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 43001716,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"feature_conversion\": \"G>A\",\n",
    "            \"feature_strand\": \"+\",\n",
    "            \"feature_name\": \"RPL27\"\n",
    "        }\n",
    "    },\n",
    "    \"same_pos_dif_reads_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 83199872,\n",
    "            \"count\": 9,\n",
    "            \"coverage\": 9,\n",
    "            \"conversion\": \"C>G\",\n",
    "            \"feature_conversion\": \"C>G\",\n",
    "            \"feature_strand\": \"+\",\n",
    "            \"feature_name\": \"AC139099.2\"\n",
    "        }\n",
    "    },\n",
    "    \"tax1bp3_chr17_3665556_read_test\": {\n",
    "        \"folder\": \"strandedness_tests\",\n",
    "        \"expectations\": {\n",
    "            \"contig\": \"chr17\",\n",
    "            \"position\": 3665556,\n",
    "            \"num_rows\": 3,\n",
    "            \"count\": 1,\n",
    "            \"coverage\": 1,\n",
    "            \"conversion\": \"G>A\",\n",
    "            \"feature_conversion\": \"C>T\",\n",
    "            \"feature_strand\": \"-\",\n",
    "            \"feature_name\": \"TAX1BP3\"\n",
    "        }\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6dfed84a-89e4-433c-8b24-91ac87266fc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for only_5_cells_test\n",
      "\tExpecting: {'contig': '9', 'barcode': 'GGGACCTTCGAGCCAC-1', 'position': 3000527, 'num_rows': 1, 'count': 10, 'coverage': 12, 'conversion': 'T>C'}\n",
      "Num rows expected: 1, was 2\n",
      "Exception: count was 10\n",
      "Exception: coverage was 12.0\n",
      "Exception: conversion was T>C\n",
      "Exception: feature_conversion was nan\n",
      "Exception: feature_name was .\n",
      "\n",
      "\t >>> only_5_cells_test passed! <<<\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F1R2_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43044352, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'feature_conversion': 'C>T', 'feature_strand': '-', 'feature_name': 'BRCA1'}\n",
      "Exception: count was 1\n",
      "Exception: coverage was 1.0\n",
      "Exception: conversion was G>A\n",
      "Exception: feature_conversion was C>T\n",
      "Exception: feature_name was BRCA1\n",
      "\n",
      "\t >>> F1R2_pair_test passed! <<<\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for F2R1_end_second_in_pair_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 43001716, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'feature_conversion': 'G>A', 'feature_strand': '+', 'feature_name': 'RPL27'}\n",
      "Exception: count was 1\n",
      "Exception: coverage was 1.0\n",
      "Exception: conversion was G>A\n",
      "Exception: feature_conversion was G>A\n",
      "Exception: feature_name was RPL27\n",
      "\n",
      "\t >>> F2R1_end_second_in_pair_test passed! <<<\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for same_pos_dif_reads_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 83199872, 'count': 9, 'coverage': 9, 'conversion': 'C>G', 'feature_conversion': 'C>G', 'feature_strand': '+', 'feature_name': 'AC139099.2'}\n",
      "Exception: count was 9\n",
      "Exception: coverage was 9.0\n",
      "Exception: conversion was C>G\n",
      "Exception: feature_conversion was C>G\n",
      "Exception: feature_name was AC139099.2\n",
      "\n",
      "\t >>> same_pos_dif_reads_test passed! <<<\n",
      "\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "Checking results for tax1bp3_chr17_3665556_read_test\n",
      "\tExpecting: {'contig': 'chr17', 'position': 3665556, 'num_rows': 3, 'count': 1, 'coverage': 1, 'conversion': 'G>A', 'feature_conversion': 'C>T', 'feature_strand': '-', 'feature_name': 'TAX1BP3'}\n",
      "Exception: count was 1\n",
      "Exception: coverage was 1.0\n",
      "Exception: conversion was G>A\n",
      "Exception: feature_conversion was C>T\n",
      "Exception: feature_name was TAX1BP3\n",
      "\n",
      "\t >>> tax1bp3_chr17_3665556_read_test passed! <<<\n"
     ]
    }
   ],
   "source": [
    "for test_name, info in test_name_to_expectations.items():\n",
    "    print(\"\\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\\nChecking results for {}\".format(test_name))\n",
    "    \n",
    "    expectations = info.get(\"expectations\")\n",
    "    print(\"\\tExpecting: {}\".format(expectations))\n",
    "\n",
    "          \n",
    "    folder = info.get(\"folder\")\n",
    "    \n",
    "    contig = expectations.get(\"contig\")\n",
    "    barcode = expectations.get(\"barcode\", None)\n",
    "    \n",
    "    position = expectations.get(\"position\")\n",
    "    \n",
    "    final_filtered_site_info_annotated = \"{}/{}/final_filtered_site_info_annotated.tsv\".format(folder, test_name)\n",
    "    final_filtered_site_info_annotated_df = pd.read_csv(final_filtered_site_info_annotated, sep='\\t', index_col=0)\n",
    "\n",
    "    row_of_interest = final_filtered_site_info_annotated_df[\n",
    "        (final_filtered_site_info_annotated_df['position'] == position) &\\\n",
    "        (final_filtered_site_info_annotated_df['contig'].astype(str) == contig)\n",
    "    ]\n",
    "\n",
    "\n",
    "    if barcode:\n",
    "        row_of_interest = row_of_interest[row_of_interest.barcode == barcode]\n",
    "\n",
    "    try:\n",
    "        assert(len(row_of_interest) == expectations.get(\"num_rows\", 1))\n",
    "    except Exception as e:\n",
    "        print(\"Num rows expected: {}, was {}\".format(expectations.get(\"num_rows\", 1), len(row_of_interest)))\n",
    "        \n",
    "    for attribute in ['count', 'coverage', 'conversion', 'feature_conversion', 'feature_name']:\n",
    "        try:\n",
    "            assert(row_of_interest[attribute.iloc[0] == expectations.get(attribute)])\n",
    "        except Exception as e:\n",
    "            print(\"Exception: {} was {}\".format(attribute, row_of_interest[attribute].iloc[0]))\n",
    "            \n",
    "    print(\"\\n\\t >>> {} passed! <<<\".format(test_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b07b328-09c2-4d4e-91c5-f3f22b041648",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fa0520-9e55-4715-ae44-bce13811802a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "workhorse",
   "language": "python",
   "name": "workhorse"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
