{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "834c9266-dd67-464d-b30d-eb56aee77724",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fba3a6f1-60ab-4ec3-9f17-a59ab8d4bbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "bampath = '/projects/ps-yeolab5/ekofman/Sammi/MouseBrainEF1A_SingleCell_EPR_batch2/filtered_possorted_ms_hippo_stamp_bam/filtered_keep_xf25_possorted_genome_with_header.bam_MD.bam'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c2899e1-6cbd-48bd-b255-197b8aa1437e",
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(bampath, \"rb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f706672-3eeb-426c-b4e7-c403e506b10f",
   "metadata": {},
   "outputs": [],
   "source": [
    "complements = {\n",
    "    'A': 'T',\n",
    "    'C': 'G',\n",
    "    'G': 'C',\n",
    "    'T': 'A'\n",
    "}\n",
    "\n",
    "def reverse_complement(input_seq):\n",
    "    \"\"\"\n",
    "    Get the reverse complement of a sequence\n",
    "    \"\"\"\n",
    "    rev_input_seq = reversed(input_seq)\n",
    "    new_seq = []\n",
    "    for r in rev_input_seq:\n",
    "        new_seq.append(complements.get(r))\n",
    "        \n",
    "    return ''.join(new_seq)\n",
    "\n",
    "def get_positions_from_md_tag(md_tag):\n",
    "    \"\"\"\n",
    "    Figure out which positions are replaced, from the MD tag.\n",
    "    \"\"\"\n",
    "    md_tag_parsed = []\n",
    "    for c in md_tag:\n",
    "        try:\n",
    "            md_tag_parsed.append(str(int(c)))\n",
    "        except Exception as e:\n",
    "            md_tag_parsed.append('-')\n",
    "\n",
    "    positions = []\n",
    "\n",
    "    position_splitters = [int(i) for i in ''.join(md_tag_parsed).split('-')]\n",
    "    for s in position_splitters:\n",
    "        if len(positions) == 0:\n",
    "            positions.append(s)\n",
    "        else:\n",
    "            positions.append(positions[-1] + s + 1)\n",
    "    return positions\n",
    "\n",
    "\n",
    "def incorporate_insertions(aligned_sequence, cigar_tuples):\n",
    "    \"\"\"\n",
    "    Update an aligned sequence to reflect any insertions (take away those positions) such\n",
    "    that it can be better compared base-to-base to a reference sequence.\n",
    "    \"\"\"\n",
    "    new_seq = ''\n",
    "    \n",
    "    current_pos = 0\n",
    "    for mod, pos in cigar_tuples:\n",
    "        if mod == 0:\n",
    "            new_seq += aligned_sequence[current_pos:current_pos+pos]\n",
    "        current_pos += pos\n",
    "    return new_seq\n",
    "\n",
    "\n",
    "def incorporate_replaced_pos_info(aligned_seq, positions_replaced):\n",
    "    \"\"\"\n",
    "    Return the aligned sequence string, with specified positions indicated as upper case\n",
    "    and others as lower case. Also returns the bases at these positions themselves.\n",
    "    \"\"\"\n",
    "    def upper(x): return x.upper()\n",
    "    def lower(x): return x.lower()\n",
    "    \n",
    "    differences_function = upper\n",
    "    others_function = lower\n",
    "    \n",
    "    indicated_aligned_seq = []\n",
    "    bases_at_pos = []\n",
    "    for i, a in enumerate(aligned_seq):\n",
    "        if i in positions_replaced:\n",
    "            indicated_aligned_seq.append(differences_function(a))\n",
    "            bases_at_pos.append(a.upper())\n",
    "        else:\n",
    "            indicated_aligned_seq.append(others_function(a))\n",
    "    return indicated_aligned_seq, bases_at_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5481bde1-a0e7-43b6-8699-1043a6d7cae3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MD tag 14G3G25G2G26C24\n",
      "18M2I81M\n",
      "Reverse? True\n",
      "Forward aligned sequence\n",
      " ttattttagttcctTaaaTttttttctggactacataagtgatgTgaAaaagaattaagtggttagaacatgtaAtttccacccatcaactatgtgtat\n",
      "Reference sequence\n",
      " ttattttagttcctGaaaGttttttctggactacataagtgatgGgaGaaagaattaagtggttagaacatgtaCtttccacccatcaactatgtgtat\n",
      "barcode CAGCACGGTTCCTAAG\n",
      "Alt\tRef\tPosition\n",
      "T \t G \t 14\n",
      "T \t G \t 18\n",
      "T \t G \t 44\n",
      "A \t G \t 47\n",
      "A \t C \t 74\n",
      "\n",
      "MD tag 6A2G0T2T5G1C23T27\n",
      "73M28S\n",
      "Reverse? True\n",
      "Forward aligned sequence\n",
      " ttcctgGaaCGttAttctgAaAtacataagtgatgggagaaagaaCtaagtggttagaacatgtactttccac\n",
      "Reference sequence\n",
      " ttcctgAaaGTttTttctgGaCtacataagtgatgggagaaagaaTtaagtggttagaacatgtactttccac\n",
      "barcode CATCAAGCAAACGTGG\n",
      "Alt\tRef\tPosition\n",
      "G \t A \t 6\n",
      "C \t G \t 9\n",
      "G \t T \t 10\n",
      "A \t T \t 13\n",
      "A \t G \t 19\n",
      "A \t C \t 21\n",
      "C \t T \t 45\n",
      "\n",
      "MD tag 3G2T12T1C7A3A29\n",
      "63M38S\n",
      "Reverse? True\n",
      "Forward aligned sequence\n",
      " ttgAagAtttgttccattcAtTatctctcCacaTatgtataagtgtaaggatgtctggctgtg\n",
      "Reference sequence\n",
      " ttgGagTtttgttccattcTtCatctctcAacaAatgtataagtgtaaggatgtctggctgtg\n",
      "barcode GTCAAACCATGACAAA\n",
      "Alt\tRef\tPosition\n",
      "A \t G \t 3\n",
      "A \t T \t 6\n",
      "A \t T \t 19\n",
      "T \t C \t 21\n",
      "C \t A \t 29\n",
      "T \t A \t 33\n",
      "\n",
      "MD tag 89A2\n",
      "1S92M8S\n",
      "Reverse? True\n",
      "Forward aligned sequence\n",
      " tccttccttccttccttccttccttccttccttccttccttccttccttcctttctttctttctttctttctttctttctttctttcggGaa\n",
      "Reference sequence\n",
      " tccttccttccttccttccttccttccttccttccttccttccttccttcctttctttctttctttctttctttctttctttctttcggAaa\n",
      "barcode ACTCCCACAACTCCCT\n",
      "Alt\tRef\tPosition\n",
      "G \t A \t 89\n",
      "\n",
      "MD tag 71\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 71\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 76\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 71\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 21A44\n",
      "12S66M23S\n",
      "Reverse? True\n",
      "Forward aligned sequence\n",
      " attttctttgattgttgtgccGatgttctctatggaatcttctgcacctgagattctctcttccat\n",
      "Reference sequence\n",
      " attttctttgattgttgtgccAatgttctctatggaatcttctgcacctgagattctctcttccat\n",
      "barcode CAAGGGAGTGATATAG\n",
      "Alt\tRef\tPosition\n",
      "G \t A \t 21\n",
      "\n",
      "MD tag 73\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 73\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 101\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 74\n",
      "\t\t\t\t\t\tMaybe no edits\n",
      "MD tag 73\n",
      "\t\t\t\t\t\tMaybe no edits\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for read in samfile.fetch('16', 3495912, 3505912):\n",
    "    md_tag = read.get_tag('MD')\n",
    "    print('MD tag', md_tag)\n",
    "\n",
    "    # Are there any replacements?\n",
    "    if not ('G' in md_tag or 'A' in md_tag or 'T' in md_tag or 'C' in md_tag):\n",
    "        print('\\t\\t\\t\\t\\t\\tMaybe no edits')\n",
    "        continue\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    positions_replaced = get_positions_from_md_tag(md_tag)\n",
    "    \n",
    "    print(read.cigarstring)\n",
    "    cigar_tuples = read.cigartuples\n",
    "\n",
    "    print('Reverse?', read.is_reverse)\n",
    "    \n",
    "    aligned_seq = reverse_complement(read.get_forward_sequence())\n",
    "    aligned_seq = incorporate_insertions(aligned_seq, cigar_tuples)\n",
    "    indicated_aligned_seq, alt_bases = incorporate_replaced_pos_info(aligned_seq, positions_replaced)\n",
    "    \n",
    "    \n",
    "    reference_seq = read.get_reference_sequence().lower()\n",
    "    indicated_referece_seq, ref_bases = incorporate_replaced_pos_info(reference_seq, positions_replaced)\n",
    "\n",
    "    \n",
    "    print('Forward aligned sequence\\n', ''.join(indicated_aligned_seq))\n",
    "    print('Reference sequence\\n', ''.join(indicated_referece_seq))\n",
    "    print('barcode', read.get_tag('CR'))\n",
    "    print(\"Alt\\tRef\\tPosition\")\n",
    "    for alt, ref, pos in zip(alt_bases, ref_bases, positions_replaced):\n",
    "        print(alt, '\\t', ref, '\\t', pos)\n",
    "        \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aba9ae6-8cb9-414f-bc4c-96498172f83f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "new_anaconda3",
   "language": "python",
   "name": "new_anaconda3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
